--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
@@ -31,8 +31,8 @@
     private Block field_174853_f;
     protected boolean field_174854_a;
     public int field_70191_b;
-    private EntityLivingBase field_70192_c;
-    private String field_85053_h;
+    public EntityLivingBase field_70192_c;
+    public String field_85053_h;
     private int field_70194_h;
     private int field_70195_i;
     public Entity field_184539_c;
