--- ../src-base/minecraft/net/minecraft/world/chunk/BlockStatePaletteRegistry.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/BlockStatePaletteRegistry.java
@@ -8,7 +8,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class BlockStatePaletteRegistry implements IBlockStatePalette
-{
+{	
     public int func_186041_a(IBlockState p_186041_1_)
     {
         int i = Block.field_176229_d.func_148747_b(p_186041_1_);
