--- ../src-base/minecraft/net/minecraft/world/chunk/Chunk.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/Chunk.java
@@ -95,28 +95,36 @@
     public Chunk(World p_i45645_1_, ChunkPrimer p_i45645_2_, int p_i45645_3_, int p_i45645_4_)
     {
         this(p_i45645_1_, p_i45645_3_, p_i45645_4_);
-        int i = 256;
+        //int i = 256;
         boolean flag = !p_i45645_1_.field_73011_w.func_177495_o();
 
-        for (int j = 0; j < 16; ++j)
-        {
-            for (int k = 0; k < 16; ++k)
+        for (int l = 0; l < 16; ++l/*int j = 0; j < 16; ++j*/)
+        {	
+        	if(p_i45645_2_.ebsIsMaybeEmpty(l)) {
+        		continue;
+        	}
+        	this.field_76652_q[l/*i1*/] = new ExtendedBlockStorage(l/*i1*/ << 4, flag);
+        	
+            for (int by = 0; by < 16; by++/*int k = 0; k < 16; ++k*/)
             {
-                for (int l = 0; l < 256; ++l)
+                for (int j = 0; j < 16; ++j)
                 {
-                    IBlockState iblockstate = p_i45645_2_.func_177856_a(j, l, k);
+                	for(int k = 0; k < 16; k++){
+                		int y = (l << 4) | by;
+                		IBlockState iblockstate = p_i45645_2_.func_177856_a(j, y, k);
 
-                    if (iblockstate.func_185904_a() != Material.field_151579_a)
-                    {
-                        int i1 = l >> 4;
+                		if (iblockstate.func_185904_a() != Material.field_151579_a)
+                		{
+                			//int i1 = l >> 4;
 
-                        if (this.field_76652_q[i1] == field_186036_a)
-                        {
-                            this.field_76652_q[i1] = new ExtendedBlockStorage(i1 << 4, flag);
-                        }
+                        	//if (this.storageArrays[l/*i1*/] == NULL_BLOCK_STORAGE)
+                        	//{
+                        	//	this.storageArrays[l/*i1*/] = new ExtendedBlockStorage(l/*i1*/ << 4, flag);
+                        	//}
 
-                        this.field_76652_q[i1].func_177484_a(j, l & 15, k, iblockstate);
-                    }
+                        	this.field_76652_q[l/*i1*/].func_177484_a(j, by, k, iblockstate);
+                		}
+                	}
                 }
             }
         }
