--- ../src-base/minecraft/org/bukkit/command/defaults/HelpCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/HelpCommand.java
@@ -165,7 +165,7 @@
 
     /**
      * Computes the Dameraur-Levenshtein Distance between two strings. Adapted
-     * from the algorithm at <a href="http://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance">Wikipedia: Damerau–Levenshtein distance</a>
+     * from the algorithm at <a href="http://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance">Wikipedia: Damerau?Levenshtein distance</a>
      *
      * @param s1 The first string being compared.
      * @param s2 The second string being compared.
