--- ../src-base/minecraft/net/minecraft/scoreboard/Scoreboard.java
+++ ../src-work/minecraft/net/minecraft/scoreboard/Scoreboard.java
@@ -19,7 +19,7 @@
     private final ScoreObjective[] field_96541_d = new ScoreObjective[19];
     private final Map<String, ScorePlayerTeam> field_96542_e = Maps.<String, ScorePlayerTeam>newHashMap();
     private final Map<String, ScorePlayerTeam> field_96540_f = Maps.<String, ScorePlayerTeam>newHashMap();
-    private static String[] field_178823_g = null;
+    private static volatile String[] field_178823_g = null;
 
     @Nullable
     public ScoreObjective func_96518_b(String p_96518_1_)
