--- ../src-base/minecraft/org/bukkit/event/player/PlayerPortalEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerPortalEvent.java
@@ -5,6 +5,8 @@
 import org.bukkit.entity.Player;
 import org.bukkit.event.HandlerList;
 
+import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
+
 /**
  * Called when a player is about to teleport because it is in contact with a
  * portal.
