--- ../src-base/minecraft/net/minecraft/entity/player/EntityPlayer.java
+++ ../src-work/minecraft/net/minecraft/entity/player/EntityPlayer.java
@@ -91,6 +91,21 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.entity.CraftItem;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.entity.Player;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.player.PlayerBedEnterEvent;
+import org.bukkit.event.player.PlayerBedLeaveEvent;
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.event.player.PlayerVelocityEvent;
+import org.bukkit.util.Vector;
+// CraftBukkit end
+
 @SuppressWarnings("incomplete-switch")
 public abstract class EntityPlayer extends EntityLivingBase
 {
@@ -107,7 +122,7 @@
     private InventoryEnderChest field_71078_a = new InventoryEnderChest();
     public Container field_71069_bz;
     public Container field_71070_bA;
-    protected FoodStats field_71100_bB = new FoodStats();
+    protected FoodStats field_71100_bB = new FoodStats(this);
     protected int field_71101_bC;
     public float field_71107_bF;
     public float field_71109_bG;
@@ -118,9 +133,9 @@
     public double field_71094_bP;
     public double field_71095_bQ;
     public double field_71085_bR;
-    protected boolean field_71083_bS;
+    public boolean field_71083_bS;
     public BlockPos field_71081_bT;
-    private int field_71076_b;
+    public int field_71076_b;
     public float field_71079_bU;
     @SideOnly(Side.CLIENT)
     public float field_71082_cx;
@@ -141,6 +156,17 @@
     private ItemStack field_184831_bT = null;
     private final CooldownTracker field_184832_bU = this.func_184815_l();
     public EntityFishHook field_71104_cf;
+    
+    // CraftBukkit start
+    public boolean fauxSleeping;
+    public String spawnWorld = "";
+    public int oldLevel = -1;
+    
+    @Override
+    public CraftHumanEntity getBukkitEntity() {
+    	return (CraftHumanEntity) super.getBukkitEntity();
+    }
+    // CraftBukkit end
 
     protected CooldownTracker func_184815_l()
     {
