--- ../src-base/minecraft/net/minecraft/entity/player/InventoryPlayer.java
+++ ../src-work/minecraft/net/minecraft/entity/player/InventoryPlayer.java
@@ -1,7 +1,16 @@
 package net.minecraft.entity.player;
 
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
+
 import javax.annotation.Nullable;
+
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+
+import fewizz.svarka.inventory.IBukkitInventory;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.crash.CrashReport;
 import net.minecraft.crash.CrashReportCategory;
@@ -21,7 +30,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class InventoryPlayer implements IInventory
+public class InventoryPlayer implements IBukkitInventory
 {
     public final ItemStack[] field_70462_a = new ItemStack[36];
     public final ItemStack[] field_70460_b = new ItemStack[4];
@@ -32,6 +41,48 @@
     private ItemStack field_70457_g;
     public boolean field_70459_e;
 
+    // Craftbukkit start
+    public List transaction = new ArrayList();
+    private int maxStack = 64;
+
+
+    public ItemStack[] getContents() {
+       ItemStack[] combined = new ItemStack[this.field_70462_a.length + this.field_70460_b.length + this.field_184439_c.length];
+       System.arraycopy(this.field_70462_a, 0, combined, 0, this.field_70462_a.length);
+       System.arraycopy(this.field_70460_b, 0, combined, this.field_70462_a.length, this.field_70460_b.length);
+       System.arraycopy(this.field_184439_c, 0, combined, this.field_70462_a.length + this.field_70460_b.length, this.field_184439_c.length);
+       return combined;
+    }
+
+    public ItemStack[] getArmorContents() {
+       return this.field_70460_b;
+    }
+
+    public void onOpen(CraftHumanEntity who) {
+       this.transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+       this.transaction.remove(who);
+    }
+
+    public List getViewers() {
+       return this.transaction;
+    }
+
+    public InventoryHolder getOwner() {
+       return this.field_70458_d.getBukkitEntity();
+    }
+
+    public void setMaxStackSize(int size) {
+       this.maxStack = size;
+    }
+
+    public Location getLocation() {
+       return this.field_70458_d.getBukkitEntity().getLocation();
+    }
+    // Craftbukkit end
+    
     public InventoryPlayer(EntityPlayer p_i1750_1_)
     {
         this.field_184440_g = new ItemStack[][] {this.field_70462_a, this.field_70460_b, this.field_184439_c};
