--- ../src-base/minecraft/net/minecraft/block/BlockJukebox.java
+++ ../src-work/minecraft/net/minecraft/block/BlockJukebox.java
@@ -62,7 +62,7 @@
         }
     }
 
-    private void func_180678_e(World p_180678_1_, BlockPos p_180678_2_, IBlockState p_180678_3_)
+    public void func_180678_e(World p_180678_1_, BlockPos p_180678_2_, IBlockState p_180678_3_)
     {
         if (!p_180678_1_.field_72995_K)
         {
