--- ../src-base/minecraft/net/minecraft/world/storage/SaveHandler.java
+++ ../src-work/minecraft/net/minecraft/world/storage/SaveHandler.java
@@ -20,6 +20,14 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import java.util.UUID;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+
+// CraftBukkit start
+import java.util.UUID;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+// CraftBukkit end
+
 public class SaveHandler implements ISaveHandler, IPlayerFileData
 {
     private static final Logger field_151478_a = LogManager.getLogger();
@@ -30,6 +38,7 @@
     private final String field_75767_f;
     private final TemplateManager field_186342_h;
     protected final DataFixer field_186341_a;
+    private UUID uuid = null; // CraftBukkit
 
     public SaveHandler(File p_i46648_1_, String p_i46648_2_, boolean p_i46648_3_, DataFixer p_i46648_4_)
     {
@@ -219,12 +228,38 @@
 
         if (nbttagcompound != null)
         {
+        	// CraftBukkit start
+        	if (p_75752_1_ instanceof EntityPlayer) {
+        		CraftPlayer cbplayer = (CraftPlayer) p_75752_1_.getBukkitEntity();
+        		// Only update first played if it is older than the one we have
+        		long modified = new File(this.field_75771_c, p_75752_1_.func_110124_au().toString() + ".dat").lastModified();
+        		if (modified < cbplayer.getFirstPlayed()) {
+        			cbplayer.setFirstPlayed(modified);
+        		}
+        	}
+        	// CraftBukkit end
             p_75752_1_.func_70020_e(this.field_186341_a.func_188257_a(FixTypes.PLAYER, nbttagcompound));
         }
 
         net.minecraftforge.event.ForgeEventFactory.firePlayerLoadingEvent(p_75752_1_, field_75771_c, p_75752_1_.func_110124_au().toString());
         return nbttagcompound;
     }
+    
+    // CraftBukkit start
+    public NBTTagCompound getPlayerData(String s) {
+    	try {
+    		File file1 = new File(this.field_75771_c, s + ".dat");
+    		
+    		if (file1.exists()) {
+    			return CompressedStreamTools.func_74796_a(new FileInputStream(file1));
+    		}
+    	} catch (Exception exception) {
+    		field_151478_a.warn("Failed to load player data for " + s);
+    	}
+   
+    	return null;
+    }
+    // CraftBukkit end
 
     public IPlayerFileData func_75756_e()
     {
@@ -264,6 +299,52 @@
     {
         return this.field_186342_h;
     }
+    
+    // CraftBukkit start
+    public UUID getUUID() {
+    	if (uuid != null) return uuid;
+    	File file1 = new File(this.field_75771_c, "uid.dat");
+    	if (file1.exists()) {
+    		DataInputStream dis = null;
+    		try {
+    			dis = new DataInputStream(new FileInputStream(file1));
+    			return uuid = new UUID(dis.readLong(), dis.readLong());
+    		} catch (IOException ex) {
+    			field_151478_a.warn("Failed to read " + file1 + ", generating new random UUID", ex);
+    		} finally {
+    			if (dis != null) {
+    				try {
+    					dis.close();
+    				} catch (IOException ex) {
+    					// NOOP
+    				}
+    			}
+    		}
+    	}
+    	uuid = UUID.randomUUID();
+    	DataOutputStream dos = null;
+    	try {
+    		dos = new DataOutputStream(new FileOutputStream(file1));
+    		dos.writeLong(uuid.getMostSignificantBits());
+    		dos.writeLong(uuid.getLeastSignificantBits());
+    	} catch (IOException ex) {
+    		field_151478_a.warn("Failed to write " + file1, ex);
+    	} finally {
+    		if (dos != null) {
+    			try {
+    				dos.close();
+    			} catch (IOException ex) {
+    				// NOOP
+    			}
+    		}
+    	}
+    	return uuid;
+    }
+    
+    public File getPlayerDir() {
+    	return field_75771_c;
+    }
+    // CraftBukkit end
 
     public NBTTagCompound getPlayerNBT(net.minecraft.entity.player.EntityPlayerMP player)
     {
