--- ../src-base/minecraft/org/bukkit/map/MapCanvas.java
+++ ../src-work/minecraft/org/bukkit/map/MapCanvas.java
@@ -71,7 +71,7 @@
     /**
      * Render text to the map using fancy formatting. Newline (\n) characters
      * will move down one line and return to the original column, and the text
-     * color can be changed using sequences such as "§12;", replacing 12 with
+     * color can be changed using sequences such as "?12;", replacing 12 with
      * the palette index of the color (see {@link MapPalette}).
      *
      * @param x The column to start rendering on.
