--- ../src-base/minecraft/org/bukkit/World.java
+++ ../src-work/minecraft/org/bukkit/World.java
@@ -1421,7 +1421,7 @@
          * @return dimension ID
          * @deprecated Magic value
          */
-        @Deprecated
+        // @Deprecated
         public int getId() {
             return id;
         }
@@ -1433,7 +1433,7 @@
          * @return The environment
          * @deprecated Magic value
          */
-        @Deprecated
+        //@Deprecated // Svarka - whyyy? =(
         public static Environment getEnvironment(int id) {
             return lookup.get(id);
         }
