--- ../src-base/minecraft/net/minecraft/tileentity/TileEntitySign.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntitySign.java
@@ -24,7 +24,7 @@
 {
     public final ITextComponent[] field_145915_a = new ITextComponent[] {new TextComponentString(""), new TextComponentString(""), new TextComponentString(""), new TextComponentString("")};
     public int field_145918_i = -1;
-    private boolean field_145916_j = true;
+    public boolean field_145916_j = true;
     private EntityPlayer field_145917_k;
     private final CommandResultStats field_174883_i = new CommandResultStats();
 
