--- ../src-base/minecraft/net/minecraft/entity/EntityLiving.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLiving.java
@@ -5,6 +5,8 @@
 import java.util.Map;
 import java.util.Random;
 import java.util.UUID;
+import java.util.logging.Level;
+
 import javax.annotation.Nullable;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.enchantment.EnchantmentHelper;
@@ -19,6 +21,7 @@
 import net.minecraft.entity.monster.EntityGhast;
 import net.minecraft.entity.monster.IMob;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Items;
 import net.minecraft.inventory.EntityEquipmentSlot;
@@ -57,7 +60,22 @@
 import net.minecraft.world.storage.loot.LootTable;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+import org.bukkit.event.entity.EntityUnleashEvent;
+import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
 
+//CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityUnleashEvent;
+import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
+// CraftBukkit end
+
 public abstract class EntityLiving extends EntityLivingBase
 {
     private static final DataParameter<Byte> field_184654_a = EntityDataManager.<Byte>func_187226_a(EntityLiving.class, DataSerializers.field_187191_a);
@@ -68,16 +86,16 @@
     protected EntityJumpHelper field_70767_i;
     private final EntityBodyHelper field_70762_j;
     protected PathNavigate field_70699_by;
-    public final EntityAITasks field_70714_bg;
-    public final EntityAITasks field_70715_bh;
+    public EntityAITasks field_70714_bg;
+    public EntityAITasks field_70715_bh;
     private EntityLivingBase field_70696_bz;
     private final EntitySenses field_70723_bA;
     private final ItemStack[] field_184656_bv = new ItemStack[2];
-    protected float[] field_82174_bp = new float[2];
+    public float[] field_82174_bp = new float[2];
     private final ItemStack[] field_184657_bw = new ItemStack[4];
-    protected float[] field_184655_bs = new float[4];
-    private boolean field_82172_bs;
-    private boolean field_82179_bU;
+    public float[] field_184655_bs = new float[4];
+    public boolean field_82172_bs;
+    public boolean field_82179_bU;
     private final Map<PathNodeType, Float> field_184658_bz = Maps.newEnumMap(PathNodeType.class);
     private ResourceLocation field_184659_bA;
     private long field_184653_bB;
@@ -103,6 +121,9 @@
         {
             this.func_184651_r();
         }
+        // CraftBukkit start - default persistance to type's persistance value
+        this.field_82179_bU = !func_70692_ba();
+        // CraftBukkit end
     }
 
     protected void func_184651_r()
@@ -169,9 +190,39 @@
 
     public void func_70624_b(@Nullable EntityLivingBase p_70624_1_)
     {
-        this.field_70696_bz = p_70624_1_;
-        net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(this, p_70624_1_);
+    	setGoalTarget(p_70624_1_, EntityTargetEvent.TargetReason.UNKNOWN, true);
     }
+    public boolean setGoalTarget(EntityLivingBase entityliving, EntityTargetEvent.TargetReason reason, boolean fireEvent) {
+    	if (this.func_70638_az() == entityliving) {
+            return false;
+        }
+        if (fireEvent) {
+            if (reason == EntityTargetEvent.TargetReason.UNKNOWN && this.func_70638_az() != null && entityliving == null) {
+                reason = (this.func_70638_az().func_70089_S() ? EntityTargetEvent.TargetReason.FORGOT_TARGET : EntityTargetEvent.TargetReason.TARGET_DIED);
+            }
+            if (reason == EntityTargetEvent.TargetReason.UNKNOWN) {
+                this.field_70170_p.getServer().getLogger().log(Level.WARNING, "Unknown target reason, please report on the issue tracker", new Exception());
+            }
+            CraftLivingEntity ctarget = null;
+            if (entityliving != null) {
+                ctarget = (CraftLivingEntity)entityliving.getBukkitEntity();
+            }
+            final EntityTargetLivingEntityEvent event = new EntityTargetLivingEntityEvent(this.getBukkitEntity(), ctarget, reason);
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return false;
+            }
+            if (event.getTarget() != null) {
+                entityliving = ((CraftLivingEntity)event.getTarget()).getHandle();
+            }
+            else {
+                entityliving = null;
+            }
+        }
+        this.field_70696_bz = entityliving;
+        net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(this, entityliving);
+        return true;
+    }
 
     public boolean func_70686_a(Class <? extends EntityLivingBase > p_70686_1_)
     {
@@ -451,14 +502,24 @@
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
-
-        if (p_70037_1_.func_150297_b("CanPickUpLoot", 1))
-        {
-            this.func_98053_h(p_70037_1_.func_74767_n("CanPickUpLoot"));
+        // CB start
+        if (p_70037_1_.func_150297_b("CanPickUpLoot", 1)) {
+            final boolean data = p_70037_1_.func_74767_n("CanPickUpLoot");
+            if (Entity.isLevelAtLeast(p_70037_1_, 1) || data) {
+                this.func_70637_d(data);
+            }
         }
+        final boolean data = p_70037_1_.func_74767_n("PersistenceRequired");
+        if (Entity.isLevelAtLeast(p_70037_1_, 1) || data) {
+            this.field_82179_bU = data;
+        }
+        //if (compound.hasKey("CanPickUpLoot", 1))
+        //{
+        //    this.setCanPickUpLoot(compound.getBoolean("CanPickUpLoot"));
+        //}
+        //this.persistenceRequired = compound.getBoolean("PersistenceRequired");
+        // CB end
 
-        this.field_82179_bU = p_70037_1_.func_74767_n("PersistenceRequired");
-
         if (p_70037_1_.func_150297_b("ArmorItems", 9))
         {
             NBTTagList nbttaglist = p_70037_1_.func_150295_c("ArmorItems", 10);
@@ -706,7 +767,7 @@
         return true;
     }
 
-    protected boolean func_70692_ba()
+    public boolean func_70692_ba()
     {
         return true;
     }
@@ -740,12 +801,12 @@
                 double d2 = entity.field_70161_v - this.field_70161_v;
                 double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                if (this.func_70692_ba() && d3 > 16384.0D)
+                if (/*this.canDespawn() && */d3 > 16384.0D)
                 {
                     this.func_70106_y();
                 }
 
-                if (this.field_70708_bq > 600 && this.field_70146_Z.nextInt(800) == 0 && d3 > 1024.0D && this.func_70692_ba())
+                if (this.field_70708_bq > 600 && this.field_70146_Z.nextInt(800) == 0 && d3 > 1024.0D/* && this.canDespawn()*/)
                 {
                     this.func_70106_y();
                 }
@@ -1215,11 +1276,23 @@
     {
         if (this.func_110167_bD() && this.func_110166_bE() == p_184230_1_)
         {
+        	// CraftBukkit start - fire PlayerUnleashEntityEvent
+        	if (CraftEventFactory.callPlayerUnleashEntityEvent(this, p_184230_1_).isCancelled()) {
+        		((EntityPlayerMP) p_184230_1_).field_71135_a.func_147359_a(new SPacketEntityAttach(this, this.func_110166_bE()));
+        		return false;
+        	}
+        	// CraftBukkit end
             this.func_110160_i(true, !p_184230_1_.field_71075_bZ.field_75098_d);
             return true;
         }
         else if (p_184230_2_ != null && p_184230_2_.func_77973_b() == Items.field_151058_ca && this.func_184652_a(p_184230_1_))
         {
+        	// CraftBukkit start - fire PlayerLeashEntityEvent
+        	if (CraftEventFactory.callPlayerLeashEntityEvent(this, p_184230_1_, p_184230_1_).isCancelled()) {
+        		((EntityPlayerMP) p_184230_1_).field_71135_a.func_147359_a(new SPacketEntityAttach(this, this.func_110166_bE()));
+        		return false;
+        	}
+        	// CraftBukkit end
             this.func_110162_b(p_184230_1_, true);
             --p_184230_2_.field_77994_a;
             return true;
@@ -1246,11 +1319,13 @@
         {
             if (!this.func_70089_S())
             {
+            	this.field_70170_p.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.PLAYER_UNLEASH)); // CraftBukkit
                 this.func_110160_i(true, true);
             }
 
             if (this.field_110168_bw == null || this.field_110168_bw.field_70128_L)
             {
+            	this.field_70170_p.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.HOLDER_GONE)); // CraftBukkit
                 this.func_110160_i(true, true);
             }
         }
@@ -1265,7 +1340,9 @@
 
             if (!this.field_70170_p.field_72995_K && p_110160_2_)
             {
+            	this.forceDrops = true; // CraftBukkit
                 this.func_145779_a(Items.field_151058_ca, 1);
+                this.forceDrops = false; // CraftBukkit
             }
 
             if (!this.field_70170_p.field_72995_K && p_110160_1_ && this.field_70170_p instanceof WorldServer)
@@ -1349,6 +1426,7 @@
             }
             else
             {
+            	this.field_70170_p.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.UNKNOWN)); // CraftBukkit
                 this.func_110160_i(false, true);
             }
         }
