--- ../src-base/minecraft/net/minecraft/entity/EntityAreaEffectCloud.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityAreaEffectCloud.java
@@ -24,6 +24,9 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
 
+import org.bukkit.craftbukkit.entity.CraftLivingEntity; // CraftBukkit
+import org.bukkit.entity.LivingEntity; // CraftBukkit
+
 public class EntityAreaEffectCloud extends Entity
 {
     private static final DataParameter<Float> field_184498_a = EntityDataManager.<Float>func_187226_a(EntityAreaEffectCloud.class, DataSerializers.field_187193_c);
@@ -31,15 +34,15 @@
     private static final DataParameter<Boolean> field_184500_c = EntityDataManager.<Boolean>func_187226_a(EntityAreaEffectCloud.class, DataSerializers.field_187198_h);
     private static final DataParameter<Integer> field_184501_d = EntityDataManager.<Integer>func_187226_a(EntityAreaEffectCloud.class, DataSerializers.field_187192_b);
     private PotionType field_184502_e;
-    private final List<PotionEffect> field_184503_f;
+    public final List<PotionEffect> field_184503_f;
     private final Map<Entity, Integer> field_184504_g;
     private int field_184505_h;
-    private int field_184506_as;
-    private int field_184507_at;
+    public int field_184506_as;
+    public int field_184507_at;
     private boolean field_184508_au;
-    private int field_184509_av;
-    private float field_184510_aw;
-    private float field_184511_ax;
+    public int field_184509_av;
+    public float field_184510_aw;
+    public float field_184511_ax;
     private EntityLivingBase field_184512_ay;
     private UUID field_184513_az;
 
@@ -116,6 +119,22 @@
             this.func_184212_Q().func_187227_b(field_184499_b, Integer.valueOf(PotionUtils.func_185181_a(PotionUtils.func_185186_a(this.field_184502_e, this.field_184503_f))));
         }
     }
+    
+    // CraftBukkit start accessor methods
+    public void refreshEffects() {
+    	if (!this.field_184508_au) {
+    		this.func_184212_Q().func_187227_b(field_184499_b, Integer.valueOf(PotionUtils.func_185181_a(PotionUtils.func_185186_a(this.field_184502_e, this.field_184503_f)))); // PAIL: rename
+    	}
+    }
+    
+    public String getType() {
+    	return ((ResourceLocation) PotionType.field_185176_a.func_177774_c(this.field_184502_e)).toString(); // PAIL: rename
+    }
+    
+    public void setType(String string) {
+    	func_184484_a((PotionType)PotionType.field_185176_a.func_82594_a(new ResourceLocation(string))); // PAIL: rename
+    }
+    // CraftBukkit end
 
     public int func_184492_k()
     {
