--- ../src-base/minecraft/net/minecraft/entity/Entity.java
+++ ../src-work/minecraft/net/minecraft/entity/Entity.java
@@ -77,15 +77,56 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.bukkit.craftbukkit.entity.CraftEntity;
+
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.Server;
+import org.bukkit.TravelAgent;
+import org.bukkit.block.BlockFace;
+import org.bukkit.entity.Hanging;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Vehicle;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.hanging.HangingBreakByEntityEvent;
+import org.bukkit.event.vehicle.VehicleBlockCollisionEvent;
+import org.bukkit.event.vehicle.VehicleEnterEvent;
+import org.bukkit.event.vehicle.VehicleExitEvent;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityCombustEvent;
+import org.bukkit.event.entity.EntityPortalEvent;
+import org.bukkit.plugin.PluginManager;
+// CraftBukkit end
+
 public abstract class Entity implements ICommandSender, net.minecraftforge.common.capabilities.ICapabilitySerializable<NBTTagCompound>
 {
+	// CraftBukkit start
+	private static final int CURRENT_LEVEL = 2;
+	static boolean isLevelAtLeast(NBTTagCompound tag, int level) {
+		return tag.func_74764_b("Bukkit.updateLevel") && tag.func_74762_e("Bukkit.updateLevel") >= level;
+	}
+	
+	protected CraftEntity bukkitEntity;
+	
+	public CraftEntity getBukkitEntity() {
+		if (bukkitEntity == null) {
+			bukkitEntity = CraftEntity.getEntity(field_70170_p.getServer(), this);
+		}
+		return bukkitEntity;
+	}
+	// CraftBukikt end
+	
     private static final Logger field_184243_a = LogManager.getLogger();
     private static final AxisAlignedBB field_174836_a = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.0D, 0.0D, 0.0D);
     private static double field_70155_l = 1.0D;
     private static int field_70152_a;
     private int field_145783_c;
     public boolean field_70156_m;
-    private final List<Entity> field_184244_h;
+    public final List<Entity> field_184244_h;
     protected int field_184245_j;
     private Entity field_184239_as;
     public boolean field_98038_p;
@@ -128,7 +169,7 @@
     protected Random field_70146_Z;
     public int field_70173_aa;
     public int field_70174_ab;
-    private int field_70151_c;
+    public int field_70151_c;
     protected boolean field_70171_ac;
     public int field_70172_ad;
     protected boolean field_70148_d;
@@ -163,9 +204,13 @@
     protected String field_189513_ar;
     private final CommandResultStats field_174837_as;
     private final List<ItemStack> field_184235_aE;
-    protected boolean field_184238_ar;
+    public boolean field_184238_ar;
     private final Set<String> field_184236_aF;
     private boolean field_184237_aG;
+    
+    public boolean valid; // CraftBukkit
+    public org.bukkit.projectiles.ProjectileSource projectileSource; // CraftBukkit - For projectiles only
+    public boolean forceExplosionKnockback; // CraftBukkit - SPIGOT-949
 
     public Entity(World p_i1582_1_)
     {
@@ -294,7 +339,7 @@
     {
     }
 
-    protected void func_70105_a(float p_70105_1_, float p_70105_2_)
+    public void func_70105_a(float p_70105_1_, float p_70105_2_)
     {
         if (p_70105_1_ != this.field_70130_N || p_70105_2_ != this.field_70131_O)
         {
@@ -2102,12 +2147,12 @@
         this.func_70052_a(5, p_82142_1_);
     }
 
-    protected boolean func_70083_f(int p_70083_1_)
+    public boolean func_70083_f(int p_70083_1_)
     {
         return (((Byte)this.field_70180_af.func_187225_a(field_184240_ax)).byteValue() & 1 << p_70083_1_) != 0;
     }
 
-    protected void func_70052_a(int p_70052_1_, boolean p_70052_2_)
+    public void func_70052_a(int p_70052_1_, boolean p_70052_2_)
     {
         byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184240_ax)).byteValue();
 
