--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityHopper.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityHopper.java
@@ -1,7 +1,12 @@
 package net.minecraft.tileentity;
 
+import java.util.ArrayList;
 import java.util.List;
+
 import javax.annotation.Nullable;
+
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockChest;
 import net.minecraft.block.BlockHopper;
@@ -31,6 +36,31 @@
     private String field_145902_i;
     private int field_145901_j = -1;
 
+    // Craftbukkit start
+    public List transaction = new ArrayList();
+    private int maxStack = 64;
+
+    public ItemStack[] getContents() {
+       return this.field_145900_a;
+    }
+
+    public void onOpen(CraftHumanEntity who) {
+       this.transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+       this.transaction.remove(who);
+    }
+
+    public List getViewers() {
+       return this.transaction;
+    }
+
+    public void setMaxStackSize(int size) {
+       this.maxStack = size;
+    }
+    // Craftbukkit end
+    
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         super.func_145839_a(p_145839_1_);
