--- ../src-base/minecraft/net/minecraft/server/management/PlayerList.java
+++ ../src-work/minecraft/net/minecraft/server/management/PlayerList.java
@@ -8,6 +8,8 @@
 import java.io.File;
 import java.net.SocketAddress;
 import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -43,6 +45,7 @@
 import net.minecraft.scoreboard.ServerScoreboard;
 import net.minecraft.scoreboard.Team;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.network.NetHandlerLoginServer;
 import net.minecraft.stats.StatList;
 import net.minecraft.stats.StatisticsManagerServer;
 import net.minecraft.util.datafix.FixTypes;
@@ -51,6 +54,7 @@
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.TextComponentTranslation;
 import net.minecraft.util.text.TextFormatting;
+import net.minecraft.util.text.translation.I18n;
 import net.minecraft.world.GameType;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
@@ -64,7 +68,49 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.TravelAgent;
+import org.bukkit.WeatherType;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.CraftTravelAgent;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.chunkio.ChunkIOExecutor;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerChangedWorldEvent;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.event.player.PlayerLoginEvent;
+import org.bukkit.event.player.PlayerPortalEvent;
+import org.bukkit.event.player.PlayerQuitEvent;
+import org.bukkit.event.player.PlayerRespawnEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
+import org.bukkit.util.Vector;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.CraftTravelAgent;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.chunkio.ChunkIOExecutor;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.TravelAgent;
+import org.bukkit.WeatherType;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerChangedWorldEvent;
+import org.bukkit.event.player.PlayerPortalEvent;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.event.player.PlayerLoginEvent;
+import org.bukkit.event.player.PlayerQuitEvent;
+import org.bukkit.event.player.PlayerRespawnEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
+import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
+import org.bukkit.util.Vector;
+// CraftBukkit end
+
 public abstract class PlayerList
 {
     public static final File field_152613_a = new File("banned-players.json");
@@ -74,23 +120,29 @@
     private static final Logger field_148546_d = LogManager.getLogger();
     private static final SimpleDateFormat field_72403_e = new SimpleDateFormat("yyyy-MM-dd \'at\' HH:mm:ss z");
     private final MinecraftServer field_72400_f;
-    private final List<EntityPlayerMP> field_72404_b = Lists.<EntityPlayerMP>newArrayList();
+    public final List<EntityPlayerMP> field_72404_b = new java.util.concurrent.CopyOnWriteArrayList<EntityPlayerMP>(); //CB - concurrent
     private final Map<UUID, EntityPlayerMP> field_177454_f = Maps.<UUID, EntityPlayerMP>newHashMap();
     private final UserListBans field_72401_g;
     private final UserListIPBans field_72413_h;
     private final UserListOps field_72414_i;
     private final UserListWhitelist field_72411_j;
     private final Map<UUID, StatisticsManagerServer> field_148547_k;
-    private IPlayerFileData field_72412_k;
+    public IPlayerFileData field_72412_k;
     private boolean field_72409_l;
     protected int field_72405_c;
     private int field_72402_d;
     private GameType field_72410_m;
     private boolean field_72407_n;
     private int field_72408_o;
+    // CraftBukkit start
+    private CraftServer cserver;
 
     public PlayerList(MinecraftServer p_i1500_1_)
     {
+    	this.cserver = p_i1500_1_.server = new CraftServer(p_i1500_1_, this);
+    	p_i1500_1_.console = org.bukkit.craftbukkit.command.ColouredConsoleSender.getInstance();
+    	p_i1500_1_.reader.addCompleter(new org.bukkit.craftbukkit.command.ConsoleCommandCompleter(p_i1500_1_.server));
+    	// CraftBukkit end
         this.field_72401_g = new UserListBans(field_152613_a);
         this.field_72413_h = new UserListIPBans(field_152614_b);
         this.field_72414_i = new UserListOps(field_152615_c);
@@ -130,13 +182,15 @@
             s1 = p_72355_1_.func_74430_c().toString();
         }
 
-        field_148546_d.info("{}[{}] logged in with entity id {} at ({}, {}, {})", new Object[] {p_72355_2_.func_70005_c_(), s1, Integer.valueOf(p_72355_2_.func_145782_y()), Double.valueOf(p_72355_2_.field_70165_t), Double.valueOf(p_72355_2_.field_70163_u), Double.valueOf(p_72355_2_.field_70161_v)});
+        // CraftBukkit - Moved message to after join
+        //LOG.info("{}[{}] logged in with entity id {} at ({}, {}, {})", new Object[] {playerIn.getName(), s1, Integer.valueOf(playerIn.getEntityId()), Double.valueOf(playerIn.posX), Double.valueOf(playerIn.posY), Double.valueOf(playerIn.posZ)});
         WorldServer worldserver = this.field_72400_f.func_71218_a(p_72355_2_.field_71093_bK);
         WorldInfo worldinfo = worldserver.func_72912_H();
         BlockPos blockpos = worldserver.func_175694_M();
         this.func_72381_a(p_72355_2_, (EntityPlayerMP)null, worldserver);
         p_72355_2_.field_71135_a = nethandlerplayserver;
         nethandlerplayserver.func_147359_a(new SPacketJoinGame(p_72355_2_.func_145782_y(), p_72355_2_.field_71134_c.func_73081_b(), worldinfo.func_76093_s(), worldserver.field_73011_w.getDimension(), worldserver.func_175659_aa(), this.func_72352_l(), worldinfo.func_76067_t(), worldserver.func_82736_K().func_82766_b("reducedDebugInfo")));
+        p_72355_2_.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
         nethandlerplayserver.func_147359_a(new SPacketCustomPayload("MC|Brand", (new PacketBuffer(Unpooled.buffer())).func_180714_a(this.func_72365_p().getServerModName())));
         nethandlerplayserver.func_147359_a(new SPacketServerDifficulty(worldinfo.func_176130_y(), worldinfo.func_176123_z()));
         nethandlerplayserver.func_147359_a(new SPacketSpawnPosition(blockpos));
@@ -147,20 +201,25 @@
         p_72355_2_.func_147099_x().func_150884_b(p_72355_2_);
         this.func_96456_a((ServerScoreboard)worldserver.func_96441_U(), p_72355_2_);
         this.field_72400_f.func_147132_au();
-        TextComponentTranslation textcomponenttranslation;
-
+        // CraftBukkit start - login message is handled in the event
+        //TextComponentTranslation textcomponenttranslation;
+        String joinMessage;
         if (p_72355_2_.func_70005_c_().equalsIgnoreCase(s))
         {
-            textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined", new Object[] {p_72355_2_.func_145748_c_()});
+            //textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined", new Object[] {playerIn.getDisplayName()});
+        	joinMessage = "§e" + I18n.func_74837_a("multiplayer.player.joined", p_72355_2_.func_70005_c_());
         }
         else
         {
-            textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined.renamed", new Object[] {p_72355_2_.func_145748_c_(), s});
+            //textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined.renamed", new Object[] {playerIn.getDisplayName(), s});
+        	joinMessage = "§e" + I18n.func_74837_a("multiplayer.player.joined.renamed", p_72355_2_.func_70005_c_(), s);
         }
 
-        textcomponenttranslation.func_150256_b().func_150238_a(TextFormatting.YELLOW);
-        this.func_148539_a(textcomponenttranslation);
-        this.func_72377_c(p_72355_2_);
+        //textcomponenttranslation.getStyle().setColor(TextFormatting.YELLOW);
+        //this.sendChatMsg(textcomponenttranslation);
+        this.playerLoggedIn(p_72355_2_, joinMessage);
+        worldserver = this.field_72400_f.func_71218_a((p_72355_2_.field_71093_bK));
+        // CraftBukkit end
         nethandlerplayserver.func_147364_a(p_72355_2_.field_70165_t, p_72355_2_.field_70163_u, p_72355_2_.field_70161_v, p_72355_2_.field_70177_z, p_72355_2_.field_70125_A);
         this.func_72354_b(p_72355_2_, worldserver);
 
@@ -226,9 +285,11 @@
 
         p_72355_2_.func_71116_b();
         net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedIn(p_72355_2_);
+        // CraftBukkit - Moved from above, added world
+        PlayerList.field_148546_d.info(String.valueOf(p_72355_2_.func_70005_c_()) + "[" + s1 + "] logged in with entity id " + p_72355_2_.func_145782_y() + " at ([" + p_72355_2_.field_70170_p.field_72986_A.func_76065_j() + "]" + p_72355_2_.field_70165_t + ", " + p_72355_2_.field_70163_u + ", " + p_72355_2_.field_70161_v + ")");
     }
 
-    protected void func_96456_a(ServerScoreboard p_96456_1_, EntityPlayerMP p_96456_2_)
+    public void func_96456_a(ServerScoreboard p_96456_1_, EntityPlayerMP p_96456_2_)
     {
         Set<ScoreObjective> set = Sets.<ScoreObjective>newHashSet();
 
@@ -352,21 +413,47 @@
         }
     }
 
-    public void func_72377_c(EntityPlayerMP p_72377_1_)
+    public void playerLoggedIn(EntityPlayerMP playerIn, String joinMessage) // CraftBukkit added param
     {
-        this.field_72404_b.add(p_72377_1_);
-        this.field_177454_f.put(p_72377_1_.func_110124_au(), p_72377_1_);
-        this.func_148540_a(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] {p_72377_1_}));
-        WorldServer worldserver = this.field_72400_f.func_71218_a(p_72377_1_.field_71093_bK);
+        this.field_72404_b.add(playerIn);
+        this.field_177454_f.put(playerIn.func_110124_au(), playerIn);
+        // this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] {playerIn}));  // CraftBukkit - replaced with loop below
+        WorldServer worldserver = this.field_72400_f.func_71218_a(playerIn.field_71093_bK);
 
+        // CraftBukkit start
+        final PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(this.cserver.getPlayer(playerIn), joinMessage);
+        this.cserver.getPluginManager().callEvent(playerJoinEvent);
+        joinMessage = playerJoinEvent.getJoinMessage();
+        if (joinMessage != null && joinMessage.length() > 0) {
+            ITextComponent[] fromString;
+            for (int length = (fromString = CraftChatMessage.fromString(joinMessage)).length, j = 0; j < length; ++j) {
+                final ITextComponent line = fromString[j];
+                this.field_72400_f.func_184103_al().func_148540_a(new SPacketChat(line));
+            }
+        }
+        ChunkIOExecutor.adjustPoolSize(this.func_72394_k());
+        // CraftBukkit end
+        // CraftBukkit start - sendAll above replaced with this loop
+        final SPacketPlayerListItem packet = new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] { playerIn });
         for (int i = 0; i < this.field_72404_b.size(); ++i)
         {
-            p_72377_1_.field_71135_a.func_147359_a(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] {(EntityPlayerMP)this.field_72404_b.get(i)}));
+            //playerIn.connection.sendPacket(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] {(EntityPlayerMP)this.playerEntityList.get(i)}));
+        	final EntityPlayerMP entityplayer2 = this.field_72404_b.get(i);
+            if (entityplayer2.getBukkitEntity().canSee(playerIn.getBukkitEntity())) {
+                entityplayer2.field_71135_a.func_147359_a(packet);
+            }
+            if (playerIn.getBukkitEntity().canSee(entityplayer2.getBukkitEntity())) {
+                playerIn.field_71135_a.func_147359_a(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] { entityplayer2 }));
+            }
         }
-
-        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.func_72394_k());
-        worldserver.func_72838_d(p_72377_1_);
-        this.func_72375_a(p_72377_1_, (WorldServer)null);
+        // CraftBukkit end
+        // CraftBukkit start - Only add if the player wasn't moved in the event
+        if (playerIn.field_70170_p == worldserver && !worldserver.field_73010_i.contains(playerIn)) {
+        	net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.func_72394_k());
+        	worldserver.func_72838_d(playerIn);
+        	this.func_72375_a(playerIn, (WorldServer)null);
+        }
+        // CraftBukkit end
     }
 
     public void func_72358_d(EntityPlayerMP p_72358_1_)
@@ -374,21 +461,28 @@
         p_72358_1_.func_71121_q().func_184164_w().func_72685_d(p_72358_1_);
     }
 
-    public void func_72367_e(EntityPlayerMP p_72367_1_)
+    public String playerLoggedOut(EntityPlayerMP playerIn) // CraftBukkit - return string
     {
-        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedOut(p_72367_1_);
-        WorldServer worldserver = p_72367_1_.func_71121_q();
-        p_72367_1_.func_71029_a(StatList.field_75947_j);
-        this.func_72391_b(p_72367_1_);
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedOut(playerIn);
+        WorldServer worldserver = playerIn.func_71121_q();
+        playerIn.func_71029_a(StatList.field_75947_j);
+        // CraftBukkit start - Quitting must be before we do final save of data, in case plugins need to modify it
+        CraftEventFactory.handleInventoryCloseEvent(playerIn);
+        final PlayerQuitEvent playerQuitEvent = new PlayerQuitEvent(this.cserver.getPlayer(playerIn), "§e" + playerIn.func_70005_c_() + " left the game");
+        this.cserver.getPluginManager().callEvent(playerQuitEvent);
+        playerIn.getBukkitEntity().disconnect(playerQuitEvent.getQuitMessage());
+        playerIn.func_71127_g();
+        // CraftBukkit end
+        this.func_72391_b(playerIn);
 
-        if (p_72367_1_.func_184218_aH())
+        if (playerIn.func_184218_aH())
         {
-            Entity entity = p_72367_1_.func_184208_bv();
+            Entity entity = playerIn.func_184208_bv();
 
             if (entity.func_184180_b(EntityPlayerMP.class).size() == 1)
             {
                 field_148546_d.debug("Removing player mount");
-                p_72367_1_.func_184210_p();
+                playerIn.func_184210_p();
                 worldserver.func_72973_f(entity);
 
                 for (Entity entity1 : entity.func_184182_bu())
@@ -396,31 +490,68 @@
                     worldserver.func_72973_f(entity1);
                 }
 
-                worldserver.func_72964_e(p_72367_1_.field_70176_ah, p_72367_1_.field_70164_aj).func_76630_e();
+                worldserver.func_72964_e(playerIn.field_70176_ah, playerIn.field_70164_aj).func_76630_e();
             }
         }
 
-        worldserver.func_72900_e(p_72367_1_);
-        worldserver.func_184164_w().func_72695_c(p_72367_1_);
-        this.field_72404_b.remove(p_72367_1_);
-        UUID uuid = p_72367_1_.func_110124_au();
+        worldserver.func_72900_e(playerIn);
+        worldserver.func_184164_w().func_72695_c(playerIn);
+        this.field_72404_b.remove(playerIn);
+        UUID uuid = playerIn.func_110124_au();
         EntityPlayerMP entityplayermp = (EntityPlayerMP)this.field_177454_f.get(uuid);
 
-        if (entityplayermp == p_72367_1_)
+        if (entityplayermp == playerIn)
         {
             this.field_177454_f.remove(uuid);
             this.field_148547_k.remove(uuid);
         }
         net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.func_72394_k());
-
-        this.func_148540_a(new SPacketPlayerListItem(SPacketPlayerListItem.Action.REMOVE_PLAYER, new EntityPlayerMP[] {p_72367_1_}));
+        // CraftBukkit start
+        // this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.REMOVE_PLAYER, new EntityPlayerMP[] {playerIn}));
+        final SPacketPlayerListItem packet = new SPacketPlayerListItem(SPacketPlayerListItem.Action.REMOVE_PLAYER, new EntityPlayerMP[] { playerIn });
+        for (int i = 0; i < this.field_72404_b.size(); ++i) {
+            final EntityPlayerMP entityplayer3 = this.field_72404_b.get(i);
+            if (entityplayer3.getBukkitEntity().canSee(playerIn.getBukkitEntity())) {
+                entityplayer3.field_71135_a.func_147359_a(packet);
+            }
+            else {
+                entityplayer3.getBukkitEntity().removeDisconnectingPlayer(playerIn.getBukkitEntity());
+            }
+        }
+        this.cserver.getScoreboardManager().removePlayer(playerIn.getBukkitEntity());
+        ChunkIOExecutor.adjustPoolSize(this.func_72394_k());
+        return playerQuitEvent.getQuitMessage();
+        
     }
 
-    public String func_148542_a(SocketAddress p_148542_1_, GameProfile p_148542_2_)
+    // CraftBukkit start - Whole method, SocketAddress to LoginListener, added hostname to signature, return EntityPlayerMP
+    public /*String*/EntityPlayerMP allowUserToConnect(/*SocketAddress address*/final NetHandlerLoginServer loginlistener, GameProfile profile, String hostname)
     {
-        if (this.field_72401_g.func_152702_a(p_148542_2_))
+    	// Moved from processLogin
+    	final UUID uuid = EntityPlayer.func_146094_a(profile);
+        final ArrayList<EntityPlayerMP> arraylist = Lists.newArrayList();
+        for (int i = 0; i < this.field_72404_b.size(); ++i) {
+            final EntityPlayerMP entityplayer = this.field_72404_b.get(i);
+            if (entityplayer.func_110124_au().equals(uuid)) {
+                arraylist.add(entityplayer);
+            }
+        }
+        final Iterator<EntityPlayerMP> iterator = arraylist.iterator();
+        while (iterator.hasNext()) {
+            final EntityPlayerMP entityplayer = iterator.next();
+            this.func_72391_b(entityplayer);
+            entityplayer.field_71135_a.func_147360_c("You logged in from another location");
+        }
+        // Instead of kicking then returning, we need to store the kick reason
+        // in the event, check with plugins to see if it's ok, and THEN kick
+        // depending on the outcome.
+        final SocketAddress socketaddress = loginlistener.field_147333_a.func_74430_c();
+        final EntityPlayerMP entity = new EntityPlayerMP(this.field_72400_f, this.field_72400_f.field_71305_c[0], profile, new PlayerInteractionManager(this.field_72400_f.field_71305_c[0]));
+        final Player player = entity.getBukkitEntity();
+        final PlayerLoginEvent event = new PlayerLoginEvent(player, hostname, ((java.net.InetSocketAddress)socketaddress).getAddress());
+        if (/*this.bannedPlayers.isBanned(profile)*/func_152608_h().func_152702_a(profile) && !func_152608_h().func_152683_b(profile).func_73682_e())
         {
-            UserListBansEntry userlistbansentry = (UserListBansEntry)this.field_72401_g.func_152683_b(p_148542_2_);
+            UserListBansEntry userlistbansentry = (UserListBansEntry)this.field_72401_g.func_152683_b(profile);
             String s1 = "You are banned from this server!\nReason: " + userlistbansentry.func_73686_f();
 
             if (userlistbansentry.func_73680_d() != null)
@@ -428,15 +559,16 @@
                 s1 = s1 + "\nYour ban will be removed on " + field_72403_e.format(userlistbansentry.func_73680_d());
             }
 
-            return s1;
+            //return s1;
+            event.disallow(PlayerLoginEvent.Result.KICK_BANNED, s1);
         }
-        else if (!this.func_152607_e(p_148542_2_))
+        else if (!this.func_152607_e(profile))
         {
-            return "You are not white-listed on this server!";
+        	event.disallow(PlayerLoginEvent.Result.KICK_WHITELIST, "You are not white-listed on this server!");
         }
-        else if (this.field_72413_h.func_152708_a(p_148542_1_))
+        else if (this.func_72363_f().func_152708_a(socketaddress) && !this.func_72363_f().func_152709_b(socketaddress).func_73682_e())
         {
-            UserListIPBansEntry userlistipbansentry = this.field_72413_h.func_152709_b(p_148542_1_);
+            UserListIPBansEntry userlistipbansentry = this.field_72413_h.func_152709_b(socketaddress);
             String s = "Your IP address is banned from this server!\nReason: " + userlistipbansentry.func_73686_f();
 
             if (userlistipbansentry.func_73680_d() != null)
@@ -444,30 +576,40 @@
                 s = s + "\nYour ban will be removed on " + field_72403_e.format(userlistipbansentry.func_73680_d());
             }
 
-            return s;
+            //return s;
+            event.disallow(PlayerLoginEvent.Result.KICK_BANNED, s);
         }
         else
         {
-            return this.field_72404_b.size() >= this.field_72405_c && !this.func_183023_f(p_148542_2_) ? "The server is full!" : null;
+            //return this.playerEntityList.size() >= this.maxPlayers && !this.bypassesPlayerLimit(profile) ? "The server is full!" : null;
+        	if (this.field_72404_b.size() >= this.field_72405_c && !this.func_183023_f(profile)) {
+                event.disallow(PlayerLoginEvent.Result.KICK_FULL, "The server is full");
+        	}
         }
+        this.cserver.getPluginManager().callEvent(event);
+        if (event.getResult() != PlayerLoginEvent.Result.ALLOWED) {
+        	loginlistener.func_147322_a(event.getKickMessage());
+        	return null;
+        }
+        return entity;
     }
 
-    public EntityPlayerMP func_148545_a(GameProfile p_148545_1_)
+    public EntityPlayerMP createPlayerForUser(GameProfile profile, EntityPlayerMP player) // CraftBukkit - added EntityPlayer)
     {
-        UUID uuid = EntityPlayer.func_146094_a(p_148545_1_);
+        /*UUID uuid = EntityPlayer.getUUID(profile);
         List<EntityPlayerMP> list = Lists.<EntityPlayerMP>newArrayList();
 
-        for (int i = 0; i < this.field_72404_b.size(); ++i)
+        for (int i = 0; i < this.playerEntityList.size(); ++i)
         {
-            EntityPlayerMP entityplayermp = (EntityPlayerMP)this.field_72404_b.get(i);
+            EntityPlayerMP entityplayermp = (EntityPlayerMP)this.playerEntityList.get(i);
 
-            if (entityplayermp.func_110124_au().equals(uuid))
+            if (entityplayermp.getUniqueID().equals(uuid))
             {
                 list.add(entityplayermp);
             }
         }
 
-        EntityPlayerMP entityplayermp2 = (EntityPlayerMP)this.field_177454_f.get(p_148545_1_.getId());
+        EntityPlayerMP entityplayermp2 = (EntityPlayerMP)this.uuidToPlayerMap.get(profile.getId());
 
         if (entityplayermp2 != null && !list.contains(entityplayermp2))
         {
@@ -476,108 +618,217 @@
 
         for (EntityPlayerMP entityplayermp1 : list)
         {
-            entityplayermp1.field_71135_a.func_147360_c("You logged in from another location");
+            entityplayermp1.connection.kickPlayerFromServer("You logged in from another location");
         }
 
         PlayerInteractionManager playerinteractionmanager;
 
-        if (this.field_72400_f.func_71242_L())
+        if (this.mcServer.isDemo())
         {
-            playerinteractionmanager = new DemoWorldManager(this.field_72400_f.func_71218_a(0));
+            playerinteractionmanager = new DemoWorldManager(this.mcServer.worldServerForDimension(0));
         }
         else
         {
-            playerinteractionmanager = new PlayerInteractionManager(this.field_72400_f.func_71218_a(0));
+            playerinteractionmanager = new PlayerInteractionManager(this.mcServer.worldServerForDimension(0));
         }
 
-        return new EntityPlayerMP(this.field_72400_f, this.field_72400_f.func_71218_a(0), p_148545_1_, playerinteractionmanager);
+        return new EntityPlayerMP(this.mcServer, this.mcServer.worldServerForDimension(0), profile, playerinteractionmanager);*/
+    	return player;
     }
 
+    // CraftBukkit start	
     public EntityPlayerMP func_72368_a(EntityPlayerMP p_72368_1_, int p_72368_2_, boolean p_72368_3_)
     {
-        World world = field_72400_f.func_71218_a(p_72368_2_);
+    	return this.moveToWorld(p_72368_1_, p_72368_2_, p_72368_3_, null, true);
+    }
+    public EntityPlayerMP moveToWorld(EntityPlayerMP player, int dimension, final boolean conqueredEnd, Location loc, boolean avoidSuffocation) {
+        World world = field_72400_f.func_71218_a(dimension);
         if (world == null)
         {
-            p_72368_2_ = 0;
+            dimension = 0;
         }
         else if (!world.field_73011_w.func_76567_e())
         {
-            p_72368_2_ = world.field_73011_w.getRespawnDimension(p_72368_1_);
+            dimension = world.field_73011_w.getRespawnDimension(player);
         }
 
-        p_72368_1_.func_71121_q().func_73039_n().func_72787_a(p_72368_1_);
-        p_72368_1_.func_71121_q().func_73039_n().func_72790_b(p_72368_1_);
-        p_72368_1_.func_71121_q().func_184164_w().func_72695_c(p_72368_1_);
-        this.field_72404_b.remove(p_72368_1_);
-        this.field_72400_f.func_71218_a(p_72368_1_.field_71093_bK).func_72973_f(p_72368_1_);
-        BlockPos blockpos = p_72368_1_.getBedLocation(p_72368_2_);
-        boolean flag = p_72368_1_.isSpawnForced(p_72368_2_);
-        p_72368_1_.field_71093_bK = p_72368_2_;
+        player.func_71121_q().func_73039_n().func_72787_a(player);
+        //player.getServerWorld().getEntityTracker().untrackEntity(player); // CB
+        player.func_71121_q().func_184164_w().func_72695_c(player);
+        this.field_72404_b.remove(player);
+        this.field_72400_f.func_71218_a(player.field_71093_bK).func_72973_f(player);
+        BlockPos blockpos = player.getBedLocation(dimension);
+        boolean flag = player.isSpawnForced(dimension);
+        /* CB start
+        player.dimension = dimension;
         PlayerInteractionManager playerinteractionmanager;
 
-        if (this.field_72400_f.func_71242_L())
+        if (this.mcServer.isDemo())
         {
-            playerinteractionmanager = new DemoWorldManager(this.field_72400_f.func_71218_a(p_72368_1_.field_71093_bK));
+            playerinteractionmanager = new DemoWorldManager(this.mcServer.worldServerForDimension(player.dimension));
         }
         else
         {
-            playerinteractionmanager = new PlayerInteractionManager(this.field_72400_f.func_71218_a(p_72368_1_.field_71093_bK));
+            playerinteractionmanager = new PlayerInteractionManager(this.mcServer.worldServerForDimension(player.dimension));
         }
 
-        EntityPlayerMP entityplayermp = new EntityPlayerMP(this.field_72400_f, this.field_72400_f.func_71218_a(p_72368_1_.field_71093_bK), p_72368_1_.func_146103_bH(), playerinteractionmanager);
-        entityplayermp.field_71135_a = p_72368_1_.field_71135_a;
-        entityplayermp.func_71049_a(p_72368_1_, p_72368_3_);
-        entityplayermp.field_71093_bK = p_72368_2_;
-       entityplayermp.func_145769_d(p_72368_1_.func_145782_y());
-        entityplayermp.func_174817_o(p_72368_1_);
-        entityplayermp.func_184819_a(p_72368_1_.func_184591_cq());
+        EntityPlayerMP entityplayermp = new EntityPlayerMP(this.mcServer, this.mcServer.worldServerForDimension(player.dimension), player.getGameProfile(), playerinteractionmanager);
+        */
+        //EntityPlayerMP entityplayermp = player;
+        org.bukkit.World fromWorld = player.getBukkitEntity().getWorld();
+        //player.viewingCredits = false; TODO!!
+        // CraftBukkit end
+        player.field_71135_a = player.field_71135_a;
+        player.func_71049_a(player, conqueredEnd);
+        player.field_71093_bK = dimension;
+        player.func_145769_d(player.func_145782_y());
+        player.func_174817_o(player);
+        player.func_184819_a(player.func_184591_cq());
 
-        for (String s : p_72368_1_.func_184216_O())
+        for (String s : player.func_184216_O())
         {
-            entityplayermp.func_184211_a(s);
+            player.func_184211_a(s);
         }
 
-        WorldServer worldserver = this.field_72400_f.func_71218_a(p_72368_1_.field_71093_bK);
-        this.func_72381_a(entityplayermp, p_72368_1_, worldserver);
-
-        if (blockpos != null)
+        //WorldServer worldserver = this.mcServer.worldServerForDimension(player.dimension); // CB - later
+        //this.setPlayerGameTypeBasedOnOther(entityplayermp, player, worldserver);
+        // CraftBukkit start - fire PlayerRespawnEvent
+        if (loc == null)
         {
-            BlockPos blockpos1 = EntityPlayer.func_180467_a(this.field_72400_f.func_71218_a(p_72368_1_.field_71093_bK), blockpos, flag);
-
-            if (blockpos1 != null)
-            {
-                entityplayermp.func_70012_b((double)((float)blockpos1.func_177958_n() + 0.5F), (double)((float)blockpos1.func_177956_o() + 0.1F), (double)((float)blockpos1.func_177952_p() + 0.5F), 0.0F, 0.0F);
-                entityplayermp.func_180473_a(blockpos, flag);
-            }
-            else
-            {
-                entityplayermp.field_71135_a.func_147359_a(new SPacketChangeGameState(0, 0.0F));
-            }
+        	boolean isBedSpawn = false;
+        	CraftWorld cworld = (CraftWorld) this.field_72400_f.server.getWorld(player.spawnWorld);
+        	if (cworld != null && blockpos != null) {
+        		BlockPos blockpos1 = EntityPlayer.func_180467_a(this.field_72400_f.func_71218_a(player.field_71093_bK), blockpos, flag);
+        		if (blockpos1 != null)
+            	{
+        			isBedSpawn = true;
+        			loc = new Location(cworld, blockpos1.func_177958_n() + 0.5f, blockpos1.func_177956_o() + 0.1f, blockpos1.func_177952_p() + 0.5f);
+                	//entityplayermp.setLocationAndAngles((double)((float)blockpos1.getX() + 0.5F), (double)((float)blockpos1.getY() + 0.1F), (double)((float)blockpos1.getZ() + 0.5F), 0.0F, 0.0F);
+                	//entityplayermp.setSpawnPoint(blockpos, flag);
+            	}
+            	else
+            	{
+            		player.func_180473_a(blockpos, flag);
+                	player.field_71135_a.func_147359_a(new SPacketChangeGameState(0, 0.0F));
+            	}
+        	}
+        	if (loc == null) {
+        		cworld = this.field_72400_f.field_71305_c[0].getWorld();
+        		blockpos = cworld.getHandle().func_175694_M();
+        		loc = new Location(cworld, blockpos.func_177958_n() + 0.5f, blockpos.func_177956_o() + 0.1f, blockpos.func_177952_p() + 0.5f);
+        	}
+        	final Player respawnPlayer = this.cserver.getPlayer(player);
+            final PlayerRespawnEvent respawnEvent = new PlayerRespawnEvent(respawnPlayer, loc, isBedSpawn);
+            this.cserver.getPluginManager().callEvent(respawnEvent);
+            loc = respawnEvent.getRespawnLocation();
+            player.reset();
+        } else {
+        	loc.setWorld(this.field_72400_f.field_71305_c[dimension].getWorld());
         }
+        final WorldServer worldserver = ((CraftWorld)loc.getWorld()).getHandle();
+        player.func_70080_a(loc.getX(), loc.getY(), loc.getZ(), loc.getYaw(), loc.getPitch());
+        // CraftBukkit end
+        worldserver.func_72863_F().func_186025_d((int)player.field_70165_t >> 4, (int)player.field_70161_v >> 4);
 
-        worldserver.func_72863_F().func_186025_d((int)entityplayermp.field_70165_t >> 4, (int)entityplayermp.field_70161_v >> 4);
-
-        while (!worldserver.func_184144_a(entityplayermp, entityplayermp.func_174813_aQ()).isEmpty() && entityplayermp.field_70163_u < 256.0D)
+        while (avoidSuffocation && !worldserver.func_184144_a(player, player.func_174813_aQ()).isEmpty() && player.field_70163_u < 256.0D)
         {
-            entityplayermp.func_70107_b(entityplayermp.field_70165_t, entityplayermp.field_70163_u + 1.0D, entityplayermp.field_70161_v);
+            player.func_70107_b(player.field_70165_t, player.field_70163_u + 1.0D, player.field_70161_v);
         }
-
-        entityplayermp.field_71135_a.func_147359_a(new SPacketRespawn(entityplayermp.field_71093_bK, entityplayermp.field_70170_p.func_175659_aa(), entityplayermp.field_70170_p.func_72912_H().func_76067_t(), entityplayermp.field_71134_c.func_73081_b()));
+        // CraftBukkit start
+        final byte actualDimension = (byte)worldserver.getWorld().getEnvironment().getId();
+        // Force the client to refresh their chunk cache
+        if (fromWorld.getEnvironment() == worldserver.getWorld().getEnvironment()) {
+            player.field_71135_a.func_147359_a(new SPacketRespawn((byte)((actualDimension >= 0) ? -1 : 0), worldserver.func_175659_aa(), worldserver.func_72912_H().func_76067_t(), player.field_71134_c.func_73081_b()));
+        }
+        player.field_71135_a.func_147359_a(new SPacketRespawn(/*entityplayermp.dimension*/actualDimension, worldserver.func_175659_aa(), worldserver.func_72912_H().func_76067_t(), player.field_71134_c.func_73081_b()));
+        player.func_70029_a(worldserver);
+        player.field_70128_L = false;
+        player.field_71135_a.teleport(new Location(worldserver.getWorld(), player.field_70165_t, player.field_70163_u, player.field_70161_v, player.field_70177_z, player.field_70125_A));
+        player.func_70095_a(false);
         BlockPos blockpos2 = worldserver.func_175694_M();
-        entityplayermp.field_71135_a.func_147364_a(entityplayermp.field_70165_t, entityplayermp.field_70163_u, entityplayermp.field_70161_v, entityplayermp.field_70177_z, entityplayermp.field_70125_A);
-        entityplayermp.field_71135_a.func_147359_a(new SPacketSpawnPosition(blockpos2));
-        entityplayermp.field_71135_a.func_147359_a(new SPacketSetExperience(entityplayermp.field_71106_cc, entityplayermp.field_71067_cb, entityplayermp.field_71068_ca));
-        this.func_72354_b(entityplayermp, worldserver);
-        this.func_187243_f(entityplayermp);
-        worldserver.func_184164_w().func_72683_a(entityplayermp);
-        worldserver.func_72838_d(entityplayermp);
-        this.field_72404_b.add(entityplayermp);
-        this.field_177454_f.put(entityplayermp.func_110124_au(), entityplayermp);
-        entityplayermp.func_71116_b();
-        entityplayermp.func_70606_j(entityplayermp.func_110143_aJ());
-        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerRespawnEvent(entityplayermp);
-        return entityplayermp;
+        //player.connection.setPlayerLocation(player.posX, player.posY, player.posZ, player.rotationYaw, player.rotationPitch);
+        player.field_71135_a.func_147359_a(new SPacketSpawnPosition(blockpos2));
+        player.field_71135_a.func_147359_a(new SPacketSetExperience(player.field_71106_cc, player.field_71067_cb, player.field_71068_ca));
+        this.func_72354_b(player, worldserver);
+        this.func_187243_f(player);
+        //worldserver.getPlayerChunkMap().addPlayer(player);
+        //worldserver.spawnEntityInWorld(player);
+        //this.playerEntityList.add(player);
+        //this.uuidToPlayerMap.put(player.getUniqueID(), player);
+        //player.addSelfToInternalCraftingInventory();
+        if (!player.field_71135_a.isDisconnected()) {
+            worldserver.func_184164_w().func_72683_a(player);
+            worldserver.func_72838_d(player);
+            this.field_72404_b.add(player);
+            this.field_177454_f.put(player.func_110124_au(), player);
+        }
+        player.func_70606_j(player.func_110143_aJ());
+        // Added from changeDimension
+        func_72385_f(player); // Update health, etc...
+        player.func_71016_p();
+        for (final Object o1 : player.func_70651_bq()) {
+            final PotionEffect mobEffect = (PotionEffect)o1;
+            player.field_71135_a.func_147359_a(new SPacketEntityEffect(player.func_145782_y(), mobEffect));
+        }
+        if (fromWorld != loc.getWorld()) {
+            final PlayerChangedWorldEvent event = new PlayerChangedWorldEvent(player.getBukkitEntity(), fromWorld);
+            this.field_72400_f.server.getPluginManager().callEvent(event);
+        }
+        if (player.field_71135_a.isDisconnected()) {
+            this.func_72391_b(player);
+        }
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerRespawnEvent(player);
+        return player;
     }
+    public void changeDimension(final EntityPlayerMP entityplayer, final int i, final PlayerTeleportEvent.TeleportCause cause) {
+        WorldServer exitWorld = null;
+        if (entityplayer.field_71093_bK < 10) {
+            for (final WorldServer world : this.field_72400_f.field_71305_c) {
+                if (world.dimension == i) {
+                    exitWorld = world;
+                }
+            }
+        }
+        final Location enter = entityplayer.getBukkitEntity().getLocation();
+        Location exit = null;
+        boolean useTravelAgent = false;
+        if (exitWorld != null) {
+            if (cause == PlayerTeleportEvent.TeleportCause.END_PORTAL && i == 0) {
+                exit = entityplayer.getBukkitEntity().getBedSpawnLocation();
+                if (exit == null || ((CraftWorld)exit.getWorld()).getHandle().dimension != 0) {
+                    exit = exitWorld.getWorld().getSpawnLocation();
+                }
+            }
+            else {
+                exit = this.calculateTarget(enter, exitWorld);
+                useTravelAgent = true;
+            }
+        }
+        final TravelAgent agent = (TravelAgent)((exit != null) ? ((CraftWorld)exit.getWorld()).getHandle().func_85176_s() : CraftTravelAgent.DEFAULT);
+        final PlayerPortalEvent event = new PlayerPortalEvent(entityplayer.getBukkitEntity(), enter, exit, agent, cause);
+        event.useTravelAgent(useTravelAgent);
+        Bukkit.getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled() || event.getTo() == null) {
+            return;
+        }
+        exit = (event.useTravelAgent() ? event.getPortalTravelAgent().findOrCreate(event.getTo()) : event.getTo());
+        if (exit == null) {
+            return;
+        }
+        exitWorld = ((CraftWorld)exit.getWorld()).getHandle();
+        final PlayerTeleportEvent tpEvent = new PlayerTeleportEvent(entityplayer.getBukkitEntity(), enter, exit, cause);
+        Bukkit.getServer().getPluginManager().callEvent(tpEvent);
+        if (tpEvent.isCancelled() || tpEvent.getTo() == null) {
+            return;
+        }
+        final Vector velocity = entityplayer.getBukkitEntity().getVelocity();
+        exitWorld.func_85176_s().adjustExit(entityplayer, exit, velocity);
+        entityplayer.field_184851_cj = true;
+        this.moveToWorld(entityplayer, exitWorld.dimension, true, exit, false);
+        if (entityplayer.field_70159_w != velocity.getX() || entityplayer.field_70181_x != velocity.getY() || entityplayer.field_70179_y != velocity.getZ()) {
+            entityplayer.getBukkitEntity().setVelocity(velocity);
+        }
+    }
 
     public void func_187243_f(EntityPlayerMP p_187243_1_)
     {
@@ -620,8 +871,75 @@
 
     public void func_82448_a(Entity p_82448_1_, int p_82448_2_, WorldServer p_82448_3_, WorldServer p_82448_4_)
     {
-        transferEntityToWorld(p_82448_1_, p_82448_2_, p_82448_3_, p_82448_4_, p_82448_4_.func_85176_s());
+    	final Location exit = this.calculateTarget(p_82448_1_.getBukkitEntity().getLocation(), p_82448_4_);
+    	this.repositionEntity(p_82448_1_, exit, true);
+        //transferEntityToWorld(entityIn, lastDimension, oldWorldIn, toWorldIn, toWorldIn.getDefaultTeleporter());
+    }// CB
+    public Location calculateTarget(final Location enter, final World target) {
+        final WorldServer worldserver = ((CraftWorld)enter.getWorld()).getHandle();
+        WorldServer worldserver2 = target.getWorld().getHandle();
+        final int i = worldserver.dimension;
+        double y = enter.getY();
+        final float yaw = enter.getYaw();
+        final float pitch = enter.getPitch();
+        double d0 = enter.getX();
+        double d2 = enter.getZ();
+        final double d3 = 8.0;
+        if (worldserver2.dimension == -1) {
+            d0 = MathHelper.func_151237_a(d0 / d3, worldserver2.func_175723_af().func_177726_b() + 16.0, worldserver2.func_175723_af().func_177728_d() - 16.0);
+            d2 = MathHelper.func_151237_a(d2 / d3, worldserver2.func_175723_af().func_177736_c() + 16.0, worldserver2.func_175723_af().func_177733_e() - 16.0);
+        }
+        else if (worldserver2.dimension == 0) {
+            d0 = MathHelper.func_151237_a(d0 * d3, worldserver2.func_175723_af().func_177726_b() + 16.0, worldserver2.func_175723_af().func_177728_d() - 16.0);
+            d2 = MathHelper.func_151237_a(d2 * d3, worldserver2.func_175723_af().func_177736_c() + 16.0, worldserver2.func_175723_af().func_177733_e() - 16.0);
+        }
+        else {
+            BlockPos blockposition;
+            if (i == 1) {
+                worldserver2 = this.field_72400_f.field_71305_c[0];
+                blockposition = worldserver2.func_175694_M();
+            }
+            else {
+                blockposition = worldserver2.func_180504_m();
+            }
+            d0 = blockposition.func_177958_n();
+            y = blockposition.func_177956_o();
+            d2 = blockposition.func_177952_p();
+        }
+        if (i != 1) {
+            worldserver.field_72984_F.func_76320_a("placing");
+            d0 = MathHelper.func_76125_a((int)d0, -29999872, 29999872);
+            d2 = MathHelper.func_76125_a((int)d2, -29999872, 29999872);
+        }
+        return new Location(worldserver2.getWorld(), d0, y, d2, yaw, pitch);
     }
+    // CB
+    public void repositionEntity(final Entity entity, final Location exit, final boolean portal) {
+        final WorldServer worldserver = (WorldServer)entity.field_70170_p;
+        final WorldServer worldserver2 = ((CraftWorld)exit.getWorld()).getHandle();
+        final int i = worldserver.dimension;
+        entity.func_70012_b(exit.getX(), exit.getY(), exit.getZ(), exit.getYaw(), exit.getPitch());
+        if (entity.func_70089_S()) {
+            worldserver.func_72866_a(entity, false);
+        }
+        worldserver.field_72984_F.func_76319_b();
+        if (i != 1) {
+            worldserver.field_72984_F.func_76320_a("placing");
+            if (entity.func_70089_S()) {
+                if (portal) {
+                    final Vector velocity = entity.getBukkitEntity().getVelocity();
+                    worldserver2.func_85176_s().adjustExit(entity, exit, velocity);
+                    entity.func_70012_b(exit.getX(), exit.getY(), exit.getZ(), exit.getYaw(), exit.getPitch());
+                    if (entity.field_70159_w != velocity.getX() || entity.field_70181_x != velocity.getY() || entity.field_70179_y != velocity.getZ()) {
+                        entity.getBukkitEntity().setVelocity(velocity);
+                    }
+                }
+                worldserver2.func_72866_a(entity, false);
+            }
+            worldserver.field_72984_F.func_76319_b();
+        }
+        entity.func_70029_a(worldserver2);
+    }
 
     @SuppressWarnings("unused")
     public void transferEntityToWorld(Entity p_82448_1_, int p_82448_2_, WorldServer p_82448_3_, WorldServer p_82448_4_, net.minecraft.world.Teleporter teleporter)
@@ -720,7 +1038,22 @@
             ((EntityPlayerMP)this.field_72404_b.get(i)).field_71135_a.func_147359_a(p_148540_1_);
         }
     }
-
+    // CraftBukkit start - add a world/entity limited version
+    public void sendAll(final Packet packet, final EntityPlayer entityhuman) {
+        for (int i = 0; i < this.field_72404_b.size(); ++i) {
+            final EntityPlayerMP entityplayer = this.field_72404_b.get(i);
+            if (entityhuman == null || !(entityhuman instanceof EntityPlayerMP) || entityplayer.getBukkitEntity().canSee(((EntityPlayerMP)entityhuman).getBukkitEntity())) {
+                this.field_72404_b.get(i).field_71135_a.func_147359_a(packet);
+            }
+        }
+    }
+    
+    public void sendAll(final Packet packet, final World world) {
+        for (int i = 0; i < world.field_73010_i.size(); ++i) {
+            ((EntityPlayerMP)world.field_73010_i.get(i)).field_71135_a.func_147359_a(packet);
+        }
+    }
+    // CraftBukkit end
     public void func_148537_a(Packet<?> p_148537_1_, int p_148537_2_)
     {
         for (int i = 0; i < this.field_72404_b.size(); ++i)
@@ -836,12 +1169,24 @@
         int i = this.field_72400_f.func_110455_j();
         this.field_72414_i.func_152687_a(new UserListOpsEntry(p_152605_1_, this.field_72400_f.func_110455_j(), this.field_72414_i.func_183026_b(p_152605_1_)));
         this.func_187245_a(this.func_177451_a(p_152605_1_.getId()), i);
+        // CraftBukkit start
+        Player player = field_72400_f.server.getPlayer(p_152605_1_.getId());
+        if (player != null) {
+        	player.recalculatePermissions();
+        }
+        // CraftBukkit end
     }
 
     public void func_152610_b(GameProfile p_152610_1_)
     {
         this.field_72414_i.func_152684_c(p_152610_1_);
         this.func_187245_a(this.func_177451_a(p_152610_1_.getId()), 0);
+        // CraftBukkit start
+        Player player = field_72400_f.server.getPlayer(p_152610_1_.getId());
+        if (player != null) {
+        	player.recalculatePermissions();
+        }
+        // CraftBukkit end
     }
 
     private void func_187245_a(EntityPlayerMP p_187245_1_, int p_187245_2_)
@@ -896,7 +1241,11 @@
         for (int i = 0; i < this.field_72404_b.size(); ++i)
         {
             EntityPlayerMP entityplayermp = (EntityPlayerMP)this.field_72404_b.get(i);
-
+        	// CraftBukkit start - Test if player receiving packet can see the source of the packet
+        	if (entityplayermp != null && entityplayermp instanceof EntityPlayerMP && !entityplayermp.getBukkitEntity().canSee(((EntityPlayerMP) entityplayermp).getBukkitEntity())) {
+        		continue;
+        	}
+        	// CraftBukkit end
             if (entityplayermp != p_148543_1_ && entityplayermp.field_71093_bK == p_148543_10_)
             {
                 double d0 = p_148543_2_ - entityplayermp.field_70165_t;
@@ -961,16 +1310,21 @@
 
         if (p_72354_2_.func_72896_J())
         {
-            p_72354_1_.field_71135_a.func_147359_a(new SPacketChangeGameState(1, 0.0F));
-            p_72354_1_.field_71135_a.func_147359_a(new SPacketChangeGameState(7, p_72354_2_.func_72867_j(1.0F)));
-            p_72354_1_.field_71135_a.func_147359_a(new SPacketChangeGameState(8, p_72354_2_.func_72819_i(1.0F)));
+        	// CraftBukkit start - handle player weather
+        	//playerIn.connection.sendPacket(new SPacketChangeGameState(1, 0.0F));
+            //playerIn.connection.sendPacket(new SPacketChangeGameState(7, worldIn.getRainStrength(1.0F)));
+            //playerIn.connection.sendPacket(new SPacketChangeGameState(8, worldIn.getThunderStrength(1.0F)));
+        	p_72354_1_.setPlayerWeather(WeatherType.DOWNFALL, false);
+        	p_72354_1_.updateWeather(-p_72354_2_.field_73004_o, p_72354_2_.field_73004_o, -p_72354_2_.field_73017_q, p_72354_2_.field_73017_q);
+            // CraftBukkit end
         }
     }
 
     public void func_72385_f(EntityPlayerMP p_72385_1_)
     {
         p_72385_1_.func_71120_a(p_72385_1_.field_71069_bz);
-        p_72385_1_.func_71118_n();
+        //playerIn.setPlayerHealthUpdated();
+        p_72385_1_.getBukkitEntity().updateScaledHealth(); // CraftBukkit - Update scaled health on respawn and worldchange
         p_72385_1_.field_71135_a.func_147359_a(new SPacketHeldItemChange(p_72385_1_.field_71071_by.field_70461_c));
     }
 
@@ -1054,15 +1408,21 @@
     {
         for (int i = 0; i < this.field_72404_b.size(); ++i)
         {
-            ((EntityPlayerMP)this.field_72404_b.get(i)).field_71135_a.func_147360_c("Server closed");
+            ((EntityPlayerMP)this.field_72404_b.get(i)).field_71135_a.func_147360_c(this.field_72400_f.server.getShutdownMessage()); // BC
         }
     }
+    // BC
+    public void sendMessage(final ITextComponent[] iChatBaseComponents) {
+        for (final ITextComponent component : iChatBaseComponents) {
+            this.func_148544_a(component, true);
+        }
+    }
 
     public void func_148544_a(ITextComponent p_148544_1_, boolean p_148544_2_)
     {
         this.field_72400_f.func_145747_a(p_148544_1_);
         byte b0 = (byte)(p_148544_2_ ? 1 : 0);
-        this.func_148540_a(new SPacketChat(p_148544_1_, b0));
+        this.func_148540_a(new SPacketChat(CraftChatMessage.fixComponent(p_148544_1_), b0));
     }
 
     public void func_148539_a(ITextComponent p_148539_1_)
