--- ../src-base/minecraft/net/minecraft/inventory/ContainerPlayer.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerPlayer.java
@@ -1,6 +1,7 @@
 package net.minecraft.inventory;
 
 import javax.annotation.Nullable;
+
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.InventoryPlayer;
@@ -10,6 +11,17 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+
+import org.bukkit.craftbukkit.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+
+//CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+
+import fewizz.svarka.inventory.IBukkitInventory;
+// CraftBukkit end
+
 public class ContainerPlayer extends Container
 {
     private static final EntityEquipmentSlot[] field_185003_h = new EntityEquipmentSlot[] {EntityEquipmentSlot.HEAD, EntityEquipmentSlot.CHEST, EntityEquipmentSlot.LEGS, EntityEquipmentSlot.FEET};
@@ -17,11 +29,21 @@
     public IInventory field_75179_f = new InventoryCraftResult();
     public boolean field_75180_g;
     private final EntityPlayer field_82862_h;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
+    // CraftBukkit end
 
     public ContainerPlayer(final InventoryPlayer p_i1819_1_, boolean p_i1819_2_, EntityPlayer p_i1819_3_)
     {
         this.field_75180_g = p_i1819_2_;
         this.field_82862_h = p_i1819_3_;
+        // CraftBukkit start
+        this.field_75179_f = new InventoryCraftResult(); // CraftBukkit - moved to before InventoryCrafting construction
+        this.field_75181_e = new InventoryCrafting(this, 2, 2, p_i1819_1_.field_70458_d); // CraftBukkit - pass player
+        this.field_75181_e.resultInventory = this.field_75179_f; // CraftBukkit - let InventoryCrafting know about its result slot
+        this.player = p_i1819_1_; // CraftBukkit - save player
+        // CraftBukkit end
         this.func_75146_a(new SlotCrafting(p_i1819_1_.field_70458_d, this.field_75181_e, this.field_75179_f, 0, 154, 28));
 
         for (int i = 0; i < 2; ++i)
@@ -204,4 +226,17 @@
     {
         return p_94530_2_.field_75224_c != this.field_75179_f && super.func_94530_a(p_94530_1_, p_94530_2_);
     }
+    
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+    	if (bukkitEntity != null) {
+    		return bukkitEntity;
+    	}
+    
+    	CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.field_75181_e, this.field_75179_f);
+    	bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+    	return bukkitEntity;
+    }
+    // CraftBukkit end
 }
