--- ../src-base/minecraft/net/minecraft/inventory/Container.java
+++ ../src-work/minecraft/net/minecraft/inventory/Container.java
@@ -14,18 +14,45 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.inventory.InventoryView;
+
+// CraftBukkit start
+import java.util.HashMap;
+import java.util.Map;
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.Event.Result;
+import org.bukkit.event.inventory.InventoryDragEvent;
+import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.inventory.InventoryView;
+// CraftBukkit end
+
 public abstract class Container
 {
     public List<ItemStack> field_75153_a = Lists.<ItemStack>newArrayList();
     public List<Slot> field_75151_b = Lists.<Slot>newArrayList();
     public int field_75152_c;
     @SideOnly(Side.CLIENT)
-    private short field_75150_e;
+    private short field_75150_e;	
     private int field_94535_f = -1;
     private int field_94536_g;
     private final Set<Slot> field_94537_h = Sets.<Slot>newHashSet();
     protected List<IContainerListener> field_75149_d = Lists.<IContainerListener>newArrayList();
     private Set<EntityPlayer> field_75148_f = Sets.<EntityPlayer>newHashSet();
+    
+    // CraftBukkit start
+    public boolean checkReachable = true;
+    public InventoryView getBukkitView() {return null;}
+    
+    public void transferTo(Container other, org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+    	InventoryView source = this.getBukkitView(), destination = other.getBukkitView();
+    	((CraftInventory) source.getTopInventory()).getBukkitInventory().onClose(player);
+    	((CraftInventory) source.getBottomInventory()).getBukkitInventory().onClose(player);
+    	((CraftInventory) destination.getTopInventory()).getBukkitInventory().onOpen(player);
+    	((CraftInventory) destination.getBottomInventory()).getBukkitInventory().onOpen(player);
+    }
+    // CraftBukkit end
 
     protected Slot func_75146_a(Slot p_75146_1_)
     {
