--- ../src-base/minecraft/net/minecraft/server/dedicated/PropertyManager.java
+++ ../src-work/minecraft/net/minecraft/server/dedicated/PropertyManager.java
@@ -12,11 +12,15 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import joptsimple.OptionSet; // CraftBukkit
+
+import joptsimple.OptionSet;
+
 @SideOnly(Side.SERVER)
 public class PropertyManager
 {
     private static final Logger field_164440_a = LogManager.getLogger();
-    private final Properties field_73672_b = new Properties();
+    final Properties field_73672_b = new Properties();
     private final File field_73673_c;
 
     public PropertyManager(File p_i46372_1_)
@@ -58,6 +62,24 @@
             this.func_73666_a();
         }
     }
+    
+    // CraftBukkit start
+    private OptionSet options = null;
+    
+    public PropertyManager(final OptionSet options) {
+    	this((File) options.valueOf("config"));
+    	
+    	this.options = options;
+    }
+    
+    private <T> T getOverride(String name, T value) {
+    	if ((this.options != null) && (this.options.has(name))) {
+    		return (T) this.options.valueOf(name);
+    	}
+    
+    	return value;
+    }
+    // CraftBukkit end
 
     public void func_73666_a()
     {
