--- ../src-base/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
@@ -22,9 +22,9 @@
 
 public class EntityFireworkRocket extends Entity
 {
-    private static final DataParameter<Optional<ItemStack>> field_184566_a = EntityDataManager.<Optional<ItemStack>>func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187196_f);
+    public static final DataParameter<Optional<ItemStack>> field_184566_a = EntityDataManager.<Optional<ItemStack>>func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187196_f);
     private int field_92056_a;
-    private int field_92055_b;
+    public int field_92055_b;
 
     public EntityFireworkRocket(World p_i1762_1_)
     {
