--- ../src-base/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
@@ -34,7 +34,7 @@
     public int field_145812_b;
     public boolean field_145813_c = true;
     private boolean field_145808_f;
-    private boolean field_145809_g;
+    public boolean field_145809_g;
     private int field_145815_h = 40;
     private float field_145816_i = 2.0F;
     public NBTTagCompound field_145810_d;
@@ -239,7 +239,7 @@
         Block block = this.field_175132_d != null ? this.field_175132_d.func_177230_c() : Blocks.field_150350_a;
         ResourceLocation resourcelocation = (ResourceLocation)Block.field_149771_c.func_177774_c(block);
         p_70014_1_.func_74778_a("Block", resourcelocation == null ? "" : resourcelocation.toString());
-        p_70014_1_.func_74774_a("Data", (byte)block.func_176201_c(this.field_175132_d));
+        p_70014_1_.func_74774_a("Data", (byte)block.func_176201_c(this.field_175132_d == null ? Blocks.field_150350_a.func_176223_P() : field_175132_d));
         p_70014_1_.func_74768_a("Time", this.field_145812_b);
         p_70014_1_.func_74757_a("DropItem", this.field_145813_c);
         p_70014_1_.func_74757_a("HurtEntities", this.field_145809_g);
