--- ../src-base/minecraft/net/minecraft/world/Teleporter.java
+++ ../src-work/minecraft/net/minecraft/world/Teleporter.java
@@ -14,6 +14,11 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.math.MathHelper;
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.event.entity.EntityPortalExitEvent;
+import org.bukkit.util.Vector;
+// CraftBukkit end
 
 public class Teleporter
 {
@@ -27,6 +32,9 @@
         this.field_77187_a = new Random(p_i1963_1_.func_72905_C());
     }
 
+    public BlockPos findPortal(double x, double y, double z, int radius) { return null; } // CraftBukkit
+    public boolean createPortal(double x, double y, double z, int b0) {return false; } // CraftBukkit
+    
     public void func_180266_a(Entity p_180266_1_, float p_180266_2_)
     {
         if (this.field_85192_a.field_73011_w.func_186058_p().func_186068_a() != 1)
