--- ../src-base/minecraft/net/minecraftforge/client/model/pipeline/BlockInfo.java
+++ ../src-work/minecraft/net/minecraftforge/client/model/pipeline/BlockInfo.java
@@ -144,7 +144,7 @@
                     skyLight[1][x][y][z] = combine(s[1][y1][1], s[x1][y1][1], s[1][y1][z1], ty ? s[x1][y1][z1] : s[1][y1][1]);
                     blockLight[1][x][y][z] = combine(b[1][y1][1], b[x1][y1][1], b[1][y1][z1], ty ? b[x1][y1][z1] : b[1][y1][1]);
 
-                    boolean tz = translucent[1][y1][z1] || translucent[1][y1][z1];
+                    boolean tz = translucent[1][y1][z1] || translucent[x1][1][z1];
                     skyLight[2][x][y][z] = combine(s[1][1][z1], s[1][y1][z1], s[x1][1][z1], tz ? s[x1][y1][z1] : s[1][1][z1]);
                     blockLight[2][x][y][z] = combine(b[1][1][z1], b[1][y1][z1], b[x1][1][z1], tz ? b[x1][y1][z1] : b[1][1][z1]);
                 }
