--- ../src-base/minecraft/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java
@@ -800,12 +800,12 @@
     }
 
     @Override
-    public ResourceLocation getNameForObject(I p_177774_1_)
+    public ResourceLocation getNameForObject(I value)
     {
-        ResourceLocation rl = super.getNameForObjectBypass(p_177774_1_);
+        ResourceLocation rl = super.getNameForObjectBypass(value);
         if (rl == null)
         {
-            rl = activeSubstitutions.inverse().get(p_177774_1_);
+            rl = activeSubstitutions.inverse().get(value);
         }
         return rl;
     }
