--- ../src-base/minecraft/net/minecraftforge/fml/common/registry/GameRegistry.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/registry/GameRegistry.java
@@ -67,7 +67,6 @@
 import com.google.common.collect.Sets;
 import com.google.common.primitives.Ints;
 
-@SuppressWarnings({"WeakerAccess", "unused"})
 public class GameRegistry
 {
     private static Set<IWorldGenerator> worldGenerators = Sets.newHashSet();
