--- ../src-base/minecraft/net/minecraft/entity/monster/EntityCreeper.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityCreeper.java
@@ -32,6 +32,11 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+// CraftBukkit end
+
 public class EntityCreeper extends EntityMob
 {
     private static final DataParameter<Integer> field_184713_a = EntityDataManager.<Integer>func_187226_a(EntityCreeper.class, DataSerializers.field_187192_b);
@@ -230,6 +235,12 @@
         super.func_70077_a(p_70077_1_);
         this.field_70180_af.func_187227_b(field_184714_b, Boolean.valueOf(true));
     }
+    
+    // CraftBukkit start
+    public void setPowered(boolean powered) {
+    	this.field_70180_af.func_187227_b(field_184714_b, powered);
+    }
+    // CraftBukkit end
 
     protected boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_, @Nullable ItemStack p_184645_3_)
     {
