--- ../src-base/minecraft/net/minecraft/entity/monster/EntityMagmaCube.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityMagmaCube.java
@@ -43,7 +43,7 @@
         return this.field_70170_p.func_72917_a(this.func_174813_aQ(), this) && this.field_70170_p.func_184144_a(this, this.func_174813_aQ()).isEmpty() && !this.field_70170_p.func_72953_d(this.func_174813_aQ());
     }
 
-    protected void func_70799_a(int p_70799_1_)
+    public void func_70799_a(int p_70799_1_)
     {
         super.func_70799_a(p_70799_1_);
         this.func_110148_a(SharedMonsterAttributes.field_188791_g).func_111128_a((double)(p_70799_1_ * 3));
