--- ../src-base/minecraft/net/minecraft/block/Block.java
+++ ../src-work/minecraft/net/minecraft/block/Block.java
@@ -540,7 +540,8 @@
 
             for (ItemStack item : items)
             {
-                if (p_180653_1_.field_73012_v.nextFloat() <= p_180653_4_)
+            	// CraftBukkit - <= to < to allow for plugins to completely disable block drops from explosions
+                if (p_180653_1_.field_73012_v.nextFloat() < p_180653_4_)
                 {
                     func_180635_a(p_180653_1_, p_180653_2_, item);
                 }
@@ -1190,7 +1191,7 @@
      * @param metadata The Metadata of the current block
      * @return A instance of a class extending TileEntity
      */
-    public TileEntity createTileEntity(World world, IBlockState state)
+    public @Nullable TileEntity createTileEntity(World world, IBlockState state)
     {
         if (isTileProvider)
         {
@@ -1298,7 +1299,7 @@
      * @param player The player or camera entity, null in some cases.
      * @return The spawn position
      */
-    public BlockPos getBedSpawnPosition(IBlockState state, IBlockAccess world, BlockPos pos, EntityPlayer player)
+    public @Nullable BlockPos getBedSpawnPosition(IBlockState state, IBlockAccess world, BlockPos pos, EntityPlayer player)
     {
         if (world instanceof World)
             return BlockBed.func_176468_a((World)world, pos, 0);
@@ -2014,7 +2015,7 @@
     {
         @Override protected List<ItemStack> initialValue() { return new java.util.ArrayList<ItemStack>(); }
     };
-    protected List<ItemStack> captureDrops(boolean start)
+    protected @Nullable List<ItemStack> captureDrops(boolean start)
     {
         if (start)
         {
@@ -2320,7 +2321,13 @@
             }
         }
     }
-
+    
+    // CraftBukkit start
+    public int getExpDrop(World world, IBlockState data, int enchantmentLevel) {
+        return 0;
+    }
+    // CraftBukkit end
+    
     private static void func_176215_a(int p_176215_0_, ResourceLocation p_176215_1_, Block p_176215_2_)
     {
         field_149771_c.func_177775_a(p_176215_0_, p_176215_1_, p_176215_2_);
