--- ../src-base/minecraft/net/minecraft/block/BlockButton.java
+++ ../src-work/minecraft/net/minecraft/block/BlockButton.java
@@ -23,6 +23,14 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
+
+// CraftBukkit start
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public abstract class BlockButton extends BlockDirectional
 {
     public static final PropertyBool field_176584_b = PropertyBool.func_177716_a("powered");
@@ -153,6 +161,19 @@
         }
         else
         {
+        	// CraftBukkit start
+        	boolean powered = ((Boolean) p_180639_3_.func_177229_b(field_176584_b));
+        	org.bukkit.block.Block block = p_180639_1_.getWorld().getBlockAt(p_180639_2_.func_177958_n(), p_180639_2_.func_177956_o(), p_180639_2_.func_177952_p());
+        	int old = (powered) ? 15 : 0;
+        	int current = (!powered) ? 15 : 0;
+        	
+        	BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+        	p_180639_1_.getServer().getPluginManager().callEvent(eventRedstone);
+        		
+        	if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+        		return true;
+        	}
+        	// CraftBukkit end
             p_180639_1_.func_180501_a(p_180639_2_, p_180639_3_.func_177226_a(field_176584_b, Boolean.valueOf(true)), 3);
             p_180639_1_.func_175704_b(p_180639_2_, p_180639_2_);
             this.func_185615_a(p_180639_4_, p_180639_1_, p_180639_2_);
@@ -207,6 +228,16 @@
                 }
                 else
                 {
+                	// CraftBukkit start
+                	org.bukkit.block.Block block = p_180650_1_.getWorld().getBlockAt(p_180650_2_.func_177958_n(), p_180650_2_.func_177956_o(), p_180650_2_.func_177952_p());
+                		
+                	BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+                	p_180650_1_.getServer().getPluginManager().callEvent(eventRedstone);
+                		
+                	if (eventRedstone.getNewCurrent() > 0) {
+                		return;
+                	}
+                	// CraftBukkit end
                     p_180650_1_.func_175656_a(p_180650_2_, p_180650_3_.func_177226_a(field_176584_b, Boolean.valueOf(false)));
                     this.func_176582_b(p_180650_1_, p_180650_2_, (EnumFacing)p_180650_3_.func_177229_b(field_176387_N));
                     this.func_185617_b(p_180650_1_, p_180650_2_);
@@ -235,9 +266,43 @@
         List <? extends Entity > list = p_185616_2_.<Entity>func_72872_a(EntityArrow.class, p_185616_1_.func_185900_c(p_185616_2_, p_185616_3_).func_186670_a(p_185616_3_));
         boolean flag = !list.isEmpty();
         boolean flag1 = ((Boolean)p_185616_1_.func_177229_b(field_176584_b)).booleanValue();
+        
+        // CraftBukkit start - Call interact event when arrows turn on wooden buttons
+        if (flag1 != flag && flag) {
+        	org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+        	boolean allowed = false;
+        
+        	// If all of the events are cancelled block the button press, else allow
+        	for (Object object : list) {
+        		if (object != null) {
+        			EntityInteractEvent event = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+        			p_185616_2_.getServer().getPluginManager().callEvent(event);
+        
+        			if (!event.isCancelled()) {
+        				allowed = true;
+        				break;
+        			}
+        		}
+        	}
+        
+        	if (!allowed) {
+        		return;
+        	}
+        }
+        // CraftBukkit end
 
         if (flag && !flag1)
         {
+        	// CraftBukkit start
+        	org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+        		
+        	BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 0, 15);
+        	p_185616_2_.getServer().getPluginManager().callEvent(eventRedstone);
+        	
+        	if (eventRedstone.getNewCurrent() <= 0) {
+        		return;
+        	}
+        	// CraftBukkit end
             p_185616_2_.func_175656_a(p_185616_3_, p_185616_1_.func_177226_a(field_176584_b, Boolean.valueOf(true)));
             this.func_176582_b(p_185616_2_, p_185616_3_, (EnumFacing)p_185616_1_.func_177229_b(field_176387_N));
             p_185616_2_.func_175704_b(p_185616_3_, p_185616_3_);
@@ -246,6 +311,16 @@
 
         if (!flag && flag1)
         {
+        	// CraftBukkit start
+        	org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+        		
+        	BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+        	p_185616_2_.getServer().getPluginManager().callEvent(eventRedstone);
+        	
+        	if (eventRedstone.getNewCurrent() > 0) {
+        		return;
+        	}
+        	// CraftBukkit end
             p_185616_2_.func_175656_a(p_185616_3_, p_185616_1_.func_177226_a(field_176584_b, Boolean.valueOf(false)));
             this.func_176582_b(p_185616_2_, p_185616_3_, (EnumFacing)p_185616_1_.func_177229_b(field_176387_N));
             p_185616_2_.func_175704_b(p_185616_3_, p_185616_3_);
