--- ../src-base/minecraft/net/minecraft/command/server/CommandScoreboard.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandScoreboard.java
@@ -867,14 +867,14 @@
         {
             scoreboard.func_96530_a(i, scoreobjective);
 
-            if (scoreobjective != null)
-            {
+            //if (scoreobjective != null)
+            //{
                 func_152373_a(p_184919_1_, this, "commands.scoreboard.objectives.setdisplay.successSet", new Object[] {Scoreboard.func_96517_b(i), scoreobjective.func_96679_b()});
-            }
-            else
-            {
-                func_152373_a(p_184919_1_, this, "commands.scoreboard.objectives.setdisplay.successCleared", new Object[] {Scoreboard.func_96517_b(i)});
-            }
+            //}
+            //else
+            //{
+            //    notifyCommandListener(sender, this, "commands.scoreboard.objectives.setdisplay.successCleared", new Object[] {Scoreboard.getObjectiveDisplaySlot(i)});
+            //}
         }
     }
 
