--- ../src-base/minecraft/net/minecraft/command/server/CommandTeleport.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandTeleport.java
@@ -132,25 +132,31 @@
             {
                 Entity entity1 = func_184885_b(p_184881_1_, p_184881_2_, p_184881_3_[p_184881_3_.length - 1]);
 
-                if (entity1.field_70170_p != entity.field_70170_p)
+                /*if (entity1.worldObj != entity.worldObj)
                 {
                     throw new CommandException("commands.tp.notSameDimension", new Object[0]);
                 }
                 else
                 {
-                    entity.func_184210_p();
+                    entity.dismountRidingEntity();
 
                     if (entity instanceof EntityPlayerMP)
                     {
-                        ((EntityPlayerMP)entity).field_71135_a.func_147364_a(entity1.field_70165_t, entity1.field_70163_u, entity1.field_70161_v, entity1.field_70177_z, entity1.field_70125_A);
+                        ((EntityPlayerMP)entity).connection.setPlayerLocation(entity1.posX, entity1.posY, entity1.posZ, entity1.rotationYaw, entity1.rotationPitch);
                     }
                     else
                     {
-                        entity.func_70012_b(entity1.field_70165_t, entity1.field_70163_u, entity1.field_70161_v, entity1.field_70177_z, entity1.field_70125_A);
+                        entity.setLocationAndAngles(entity1.posX, entity1.posY, entity1.posZ, entity1.rotationYaw, entity1.rotationPitch);
                     }
 
-                    func_152373_a(p_184881_2_, this, "commands.tp.success", new Object[] {entity.func_70005_c_(), entity1.func_70005_c_()});
+                    notifyCommandListener(sender, this, "commands.tp.success", new Object[] {entity.getName(), entity1.getName()});
+                }*/ //Craftbukkit
+                // CraftBukkit Start
+                // Use Bukkit teleport method in all cases. It has cross dimensional handling, events
+                if (entity.getBukkitEntity().teleport((org.bukkit.entity.Entity)entity1.getBukkitEntity(), org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND)) {
+                	func_152373_a(p_184881_2_, this, "commands.tp.success", new Object[] { entity.func_70005_c_(), entity1.func_70005_c_()});
                 }
+                // CraftBukkit End
             }
         }
     }
