--- ../src-base/minecraft/org/bukkit/event/entity/EntityDamageByEntityEvent.java
+++ ../src-work/minecraft/org/bukkit/event/entity/EntityDamageByEntityEvent.java
@@ -5,6 +5,9 @@
 import com.google.common.base.Function;
 import org.bukkit.entity.Entity;
 
+import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
+import org.bukkit.event.entity.EntityDamageEvent.DamageModifier;
+
 /**
  * Called when an entity is damaged by an entity
  */
