--- ../src-base/minecraft/org/bukkit/event/entity/EntityDamageByBlockEvent.java
+++ ../src-work/minecraft/org/bukkit/event/entity/EntityDamageByBlockEvent.java
@@ -6,6 +6,9 @@
 import org.bukkit.block.Block;
 import org.bukkit.entity.Entity;
 
+import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
+import org.bukkit.event.entity.EntityDamageEvent.DamageModifier;
+
 /**
  * Called when an entity is damaged by a block
  */
