--- ../src-base/minecraft/net/minecraft/item/crafting/CraftingManager.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/CraftingManager.java
@@ -21,17 +21,23 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class CraftingManager
 {
     private static final CraftingManager field_77598_a = new CraftingManager();
-    private final List<IRecipe> field_77597_b = Lists.<IRecipe>newArrayList();
+    public List<IRecipe> field_77597_b = Lists.<IRecipe>newArrayList();
+    // CraftBukkit start
+    public IRecipe lastRecipe;
+    public org.bukkit.inventory.InventoryView lastCraftView;
+    // CraftBukkit end
 
     public static CraftingManager func_77594_a()
     {
         return field_77598_a;
     }
 
-    private CraftingManager()
+    public CraftingManager()
     {
         (new RecipesTools()).func_77586_a(this);
         (new RecipesWeapons()).func_77583_a(this);
@@ -190,14 +196,25 @@
         this.func_92103_a(new ItemStack(Blocks.field_150438_bZ), new Object[] {"I I", "ICI", " I ", 'I', Items.field_151042_j, 'C', Blocks.field_150486_ae});
         this.func_92103_a(new ItemStack(Items.field_179565_cj, 1), new Object[] {"///", " / ", "/_/", '/', Items.field_151055_y, '_', new ItemStack(Blocks.field_150333_U, 1, BlockStoneSlab.EnumType.STONE.func_176624_a())});
         this.func_92103_a(new ItemStack(Blocks.field_185764_cQ, 4), new Object[] {"/", "#", '/', Items.field_151072_bj, '#', Items.field_185162_cT});
-        Collections.sort(this.field_77597_b, new Comparator<IRecipe>()
-        {
-            public int compare(IRecipe p_compare_1_, IRecipe p_compare_2_)
-            {
-                return p_compare_1_ instanceof ShapelessRecipes && p_compare_2_ instanceof ShapedRecipes ? 1 : (p_compare_2_ instanceof ShapelessRecipes && p_compare_1_ instanceof ShapedRecipes ? -1 : (p_compare_2_.func_77570_a() < p_compare_1_.func_77570_a() ? -1 : (p_compare_2_.func_77570_a() > p_compare_1_.func_77570_a() ? 1 : 0)));
-            }
-        });
+        //Collections.sort(this.recipes, new Comparator<IRecipe>()
+        //{
+        //    public int compare(IRecipe p_compare_1_, IRecipe p_compare_2_)
+        //    {
+        //        return p_compare_1_ instanceof ShapelessRecipes && p_compare_2_ instanceof ShapedRecipes ? 1 : (p_compare_2_ instanceof ShapelessRecipes && p_compare_1_ instanceof ShapedRecipes ? -1 : (p_compare_2_.getRecipeSize() < p_compare_1_.getRecipeSize() ? -1 : (p_compare_2_.getRecipeSize() > p_compare_1_.getRecipeSize() ? 1 : 0)));
+        //    }
+        //});
+        sort();
     }
+    
+    // CraftBukkit start
+    public void sort() {
+    	Collections.sort(this.field_77597_b, new Comparator<IRecipe>() {
+    		public int compare(IRecipe p_compare_1_, IRecipe p_compare_2_)
+    		{
+    			return p_compare_1_ instanceof ShapelessRecipes && p_compare_2_ instanceof ShapedRecipes ? 1 : (p_compare_2_ instanceof ShapelessRecipes && p_compare_1_ instanceof ShapedRecipes ? -1 : (p_compare_2_.func_77570_a() < p_compare_1_.func_77570_a() ? -1 : (p_compare_2_.func_77570_a() > p_compare_1_.func_77570_a() ? 1 : 0)));
+    		}
+    	});
+    }
 
     public ShapedRecipes func_92103_a(ItemStack p_92103_1_, Object... p_92103_2_)
     {
