--- ../src-base/minecraft/net/minecraft/entity/passive/EntityHorse.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityHorse.java
@@ -57,6 +57,8 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+import org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason; // CraftBukkit
+
 public class EntityHorse extends EntityAnimal implements IInventoryChangedListener, IJumpingMount
 {
     private static final Predicate<Entity> field_110276_bu = new Predicate<Entity>()
@@ -66,7 +68,7 @@
             return p_apply_1_ instanceof EntityHorse && ((EntityHorse)p_apply_1_).func_110205_ce();
         }
     };
-    private static final IAttribute field_110271_bv = (new RangedAttribute((IAttribute)null, "horse.jumpStrength", 0.7D, 0.0D, 2.0D)).func_111117_a("Jump Strength").func_111112_a(true);
+    public static final IAttribute field_110271_bv = (new RangedAttribute((IAttribute)null, "horse.jumpStrength", 0.7D, 0.0D, 2.0D)).func_111117_a("Jump Strength").func_111112_a(true);
     private static final UUID field_184786_bD = UUID.fromString("556E1665-8B10-40C8-8F9D-CF9B1667F295");
     private static final DataParameter<Byte> field_184787_bE = EntityDataManager.<Byte>func_187226_a(EntityHorse.class, DataSerializers.field_187191_a);
     private static final DataParameter<Integer> field_184788_bF = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
@@ -84,7 +86,7 @@
     public int field_110278_bp;
     public int field_110279_bq;
     protected boolean field_110275_br;
-    private AnimalChest field_110296_bG;
+    public AnimalChest field_110296_bG;
     private boolean field_110293_bH;
     protected int field_110274_bs;
     protected float field_110277_bt;
@@ -101,6 +103,7 @@
     private String field_110286_bQ;
     private String[] field_110280_bR = new String[3];
     private boolean field_175508_bO = false;
+    public int maxDomestication = 100; // CraftBukkit - store max domestication value
 
     public EntityHorse(World p_i1685_1_)
     {
@@ -412,7 +415,7 @@
         return this.func_110261_ca() && horsetype.func_188600_f() ? 17 : 2;
     }
 
-    private void func_110226_cD()
+    public void func_110226_cD()
     {
         AnimalChest animalchest = this.field_110296_bG;
         this.field_110296_bG = new AnimalChest("HorseChest", this.func_110225_cC());
