--- ../src-base/minecraft/net/minecraft/command/CommandGameRule.java
+++ ../src-work/minecraft/net/minecraft/command/CommandGameRule.java
@@ -29,7 +29,7 @@
 
     public void func_184881_a(MinecraftServer p_184881_1_, ICommandSender p_184881_2_, String[] p_184881_3_) throws CommandException
     {
-        GameRules gamerules = this.func_184897_a(p_184881_1_);
+        GameRules gamerules = p_184881_2_.func_130014_f_().func_82736_K();//this.getOverWorldGameRules(server); // Craftbukkit
         String s = p_184881_3_.length > 0 ? p_184881_3_[0] : "";
         String s1 = p_184881_3_.length > 1 ? func_180529_a(p_184881_3_, 1) : "";
 
