--- ../src-base/minecraft/net/minecraft/command/CommandHandler.java
+++ ../src-work/minecraft/net/minecraft/command/CommandHandler.java
@@ -41,9 +41,9 @@
 
         if (icommand == null)
         {
-            TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("commands.generic.notFound", new Object[0]);
-            textcomponenttranslation.func_150256_b().func_150238_a(TextFormatting.RED);
-            p_71556_1_.func_145747_a(textcomponenttranslation);
+            //TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("commands.generic.notFound", new Object[0]);
+            //textcomponenttranslation.getStyle().setColor(TextFormatting.RED);
+            //sender.addChatMessage(textcomponenttranslation);
         }
         else if (icommand.func_184882_a(this.func_184879_a(), p_71556_1_))
         {
@@ -87,9 +87,9 @@
         }
         else
         {
-            TextComponentTranslation textcomponenttranslation1 = new TextComponentTranslation("commands.generic.permission", new Object[0]);
-            textcomponenttranslation1.func_150256_b().func_150238_a(TextFormatting.RED);
-            p_71556_1_.func_145747_a(textcomponenttranslation1);
+            //TextComponentTranslation textcomponenttranslation1 = new TextComponentTranslation("commands.generic.permission", new Object[0]);
+            //textcomponenttranslation1.getStyle().setColor(TextFormatting.RED);
+            //sender.addChatMessage(textcomponenttranslation1);
         }
 
         p_71556_1_.func_174794_a(CommandResultStats.Type.SUCCESS_COUNT, j);
