--- ../src-base/minecraft/net/minecraft/inventory/ISidedInventory.java
+++ ../src-work/minecraft/net/minecraft/inventory/ISidedInventory.java
@@ -1,9 +1,10 @@
 package net.minecraft.inventory;
 
+import fewizz.svarka.inventory.IBukkitInventory;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.EnumFacing;
 
-public interface ISidedInventory extends IInventory
+public interface ISidedInventory extends IBukkitInventory
 {
     int[] func_180463_a(EnumFacing p_180463_1_);
 
