--- ../src-base/minecraft/org/bukkit/util/noise/SimplexNoiseGenerator.java
+++ ../src-work/minecraft/org/bukkit/util/noise/SimplexNoiseGenerator.java
@@ -392,7 +392,7 @@
         // To find out which of the 24 possible simplices we're in, we need to
         // determine the magnitude ordering of x0, y0, z0 and w0.
         // The method below is a good way of finding the ordering of x,y,z,w and
-        // then find the correct traversal order for the simplex we’re in.
+        // then find the correct traversal order for the simplex we?re in.
         // First, six pair-wise comparisons are performed between each possible pair
         // of the four coordinates, and the results are used to add up binary bits
         // for an integer index.
