--- ../src-base/minecraft/net/minecraftforge/fml/client/FMLFolderResourcePack.java
+++ ../src-work/minecraft/net/minecraftforge/fml/client/FMLFolderResourcePack.java
@@ -46,9 +46,9 @@
     }
 
     @Override
-    protected boolean func_110593_b(String p_110593_1_)
+    protected boolean func_110593_b(String name)
     {
-        return super.func_110593_b(p_110593_1_);
+        return super.func_110593_b(name);
     }
     @Override
     public String func_130077_b()
