--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenBirchTree.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenBirchTree.java
@@ -103,7 +103,7 @@
                                     BlockPos blockpos = new BlockPos(i3, i2, k1);
                                     IBlockState state2 = p_180709_1_.func_180495_p(blockpos);
 
-                                    if (state2.func_177230_c().isAir(state2, p_180709_1_, blockpos) || state2.func_177230_c().isAir(state2, p_180709_1_, blockpos))
+                                    if (state2.func_177230_c().isAir(state2, p_180709_1_, blockpos))
                                     {
                                         this.func_175903_a(p_180709_1_, blockpos, field_181630_b);
                                     }
